[Security Solution] RFC for Prebuilt Rules Customization - Milestone 3 (#171856)
elastic/kibana: x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

This endpoint will need further changes, which will be detailed further down, and are out of the scope of Saved Object, migration and rule schema updates.

- [**Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal)](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/perform_rule_upgrade_route.ts)

This endpoint will require major changes to add the capability of letting users selecting a custom version of the rule with custom values for rule fields. This will be explained further below in the "Changes to upgrade `_review` and `_perform` endpoints" section.

The value of the `customized` field for the updated rule will depend on that logic as well, so its calculation will be explained in that section.

### Changes needed in `rule_monitoring` endpoints

> 

Author: I think this should do the trick but I have as a TODO to test this.

With this migration strategy for the import endpoints we can guarantee backwards compatibility for all rules, and don't need to do additional changes to the export endpoints.

The `importRules` helper method that our import endpoint handler use rely on the already discussed `createRules` and the `patchRules` - when a rule is created from scratch (no `rule_id` match), and when a `rule_id` matches and a rule is overwritten, respectively.

Given that, when importing rules, we can be creating a custom rule or a prebuilt rule, we need to calculate and appropiately pass the `isPrebuilt` flag to the `createRules` method. The `importRules` helper thus needs to be modified so:

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts)_

### Handling the `version` parameter

> In the code above, we made an additional check: if the rule to be imported is a prebuilt rule, then the `version` field will be required as part of the payload as well. Attempting to import a prebuilt rule without it will result in an error and the importing of that rule will be skipped. The `version` is necessary to be able to provide updates to the rule when a new version is available in the Prebuilt Rules package. Without it we don't know if the installed version is newer or older than the update being offered by the package.

Reviewer: Similar to the other `prebuilt` fields, this validation (prebuilt rule -> version is required) is easier to represent if `version` is a subfield of `prebuilt`. It also makes the schema semantics more straightforward, as right now `version` means different things depending on if the rule is prebuilt or not: if it's prebuilt, `version` is the prebuilt rules version. If the rule is not prebuilt, `version` is still allowed but what does it really mean compared to `revision`? 

We should weigh the cost of deprecating `version` and replacing it with `prebuilt.version` against the potential benefits. Even if we agree that using `prebuilt.version` would be the ideal state, I understand that migrating to reach that state takes time and effort and adds complexity.

-   [Customizing Prebuilt Rules](#customizing-prebuilt-rules){#toc-customizing-prebuilt-rules}
    -   [Endpoints](#endpoints){#toc-endpoints}
    -   [In the UI](#in-the-ui){#toc-in-the-ui}
-   [List of things to fix (create tickets)](#list-of-things-to-fix-create-tickets){#toc-list-of-things-to-fix-create-tickets}
    -   [Rule fields](#rule-fields){#toc-rule-fields}
-   [Updating Prebuilt Rules](#updating-prebuilt-rules){#toc-updating-prebuilt-rules}
    -   [Changes to upgrade `_review` and `_perform` endpoints](#changes-to-upgrade-_review-and-_perform-endpoints){#toc-changes-to-upgrade-_review-and-_perform-endpoints}
    -   [Rule field diff algorithms](#rule-field-diff-algorithms){#toc-rule-field-diff-algorithms}
-   [Changes in UI](#changes-in-ui){#toc-changes-in-ui}
-   [Scenarios for bulk accepting updates](#scenarios-for-bulk-accepting-updates){#toc-scenarios-for-bulk-accepting-updates}
-   [Marking rules (and their fields) as customized](#marking-rules-and-their-fields-as-customized){#toc-marking-rules-and-their-fields-as-customized}
-   [Other open questions](#other-open-questions){#toc-other-open-questions}

## Necessary rule schema changes

> In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing two new fields: `prebuilt` and `customized`. Simultaneously, we need to deprecate the existing `immutable` field, which determines if a rule is an Elastic Prebuilt Rule and whether its fields can be modified/customized.

Author: Yes. All our rules will be "mutable" after this epic, so the field name `immutable` will lose its meaning over time and become confusing, especially for consumers of the API. That's why we want to eventually deprecate it and fully replace it with `prebuilt`.

I think the change is necessary and worth the effort, but since you are right that is only a "name" change, I'm interested in your opinion as well.

Meanwhile, I'll add this explanation to the RFC in this section.

> In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing two new fields: `prebuilt` and `customized`. Simultaneously, we need to deprecate the existing `immutable` field, which determines if a rule is an Elastic Prebuilt Rule and whether its fields can be modified/customized.

Author: Actually, the `prebuilt` field should replace the `immutable` field but with one difference: the `prebuilt` field will mark if the rule is an Elastic Prebuilt Rule or not, but now all rules will be able to be customized by the user, i.e. independently of the value of `prebuilt`. That's also a semantic difference between the two fields.

To ensure backward compatibility and avoid breaking changes, we will silently deprecate the `immutable` field but keep it unchanged (i.e. non-optional).

In summary, the changes to be done to our rule schema are:

- Mark the `immutable` field as deprecated but keep unchanged for backwards compatibility.
    - Should remain a `boolean` type and non-optional (`required`)
- Add `prebuilt` field to the rule schema
    - `prebuilt` should have same type as `immutable`: `boolean`
    - Should be non-optional (`required`)
    - During migration, set the `prebuilt` value to match the current value of `immutable`. This means: prebuilt rules should be migrated with a `prebuilt` value set to `true`, while custom rules should be set with a `prebuilt` value of `false`.
- Add `customized` field to the rule schema
    - Should have type of `boolean`
    - Should be non-optional (`required`)
    - If non existing, initialize it with a default value of `false` for all rules, both custom and prebuilt. The only exception for this is when importing prebuilt rules (see below).

>     - Custom rules (`prebuilt: false`) will never use this field, and its value should always be `false`.

Reviewer: Did you consider nesting prebuilt-specific fields inside an object? When there's a relationship like this between 2 sibling fields, it's often a good candidate for making them subfields of an object. e.g. something like
```
{
  prebuilt?: {
    customized: boolean;
  }
}
```
This makes it easy later on to add additional fields that are specific to prebuilt rules. We used this pattern recently with suppression and `investigation_fields`.

Author: Hi @marshallmain we've given some thought on this with @banderror and liked your proposal, with some modifications. What do you think of the following?

```
{  
  prebuilt: {  
    isPrebuilt: boolean;
    isCustomized: boolean:
    elasticLastUpdated: Date
  }  
}  
```
First of all, we think it makes sense to have a `prebuilt` top-level object where we can group everything related to prebuilt rules. Apart from the fields I wrote above (more on that below), we though we might want to add other related fields in the future, for example a `customizations` field, which could somehow store all user's modifications to a rule's fields (for example, in case we want the user to be able to search rules by their customized field - "In which of my prebuilt rules have I modifed my index patterns field?").

But back to the current design:
- the prebuilt field shouldn't be optional, since we explicitly want to mark if a rule is prebuilt or not. So
- `isPrebuilt`: would be the first nonOptional boolean key, with the same behaviour as the `prebuilt` field as I described up to now in the RFC
- `isCustomized`: would be the second nonOptional boolean key, with the same behaviour as the `customized` field as I described up to now in the RFC
- `elasticLastUpdated`: is an optional field that

Reviewer: I think the additional fields (`customizations`, `elasticLastUpdated`, `isCustomized`) make sense, but IMO it would be simpler to leave `isPrebuilt` out entirely, make `prebuilt` optional, and use the existence of `prebuilt` to determine if a rule is prebuilt or not.
```
{  
  prebuilt?: {
    isCustomized: boolean;
    elasticLastUpdated: Date;
  }  
}
```
This makes the schema easier to represent:
```
const prebuilt = z.object({
  isCustomized: z.boolean(),
  elasticLastUpdated: z.date(),
});
type Prebuilt = z.infer<typeof prebuilt>;

const rule = z.object({
  prebuilt: prebuilt.optional(),
  field1: z.string(),
  field2: z.string(),
});
type Rule = z.infer<typeof rule>;
```

If `prebuilt` is required and some subfields take special values based on the value of `isPrebuilt`, then it's a bit more complicated to encode that `isCustomized` and other fields are only valid if `isPrebuilt` is `true`:
```
const prebuiltTrue = z.object({
  isPrebuilt: z.literal(true),
  isCustomized: z.boolean(),
  elasticLastUpdated: z.date(),
});

const prebuiltFalse = z.object({
  isPrebuilt: z.literal(false),
  isCustomized: z.literal(false), // Included per the RFC, but I'd leave this out entirely even if `isPrebuilt` does end up being necessary
});

const prebuilt = z.union([prebuiltTrue, prebuiltFalse]);
type Prebuilt = z.infer<typeof prebuilt>;

const rule = z.object({
  prebuilt,
  field1: z.string(),
  field2: z.string(),
});
type Rule = z.infer<typeof rule>;
```

To me it feels weird to include a `prebuilt` section for rules that are not prebuilt, and it seems sufficient that if `prebuilt` is missing then the rule is implicitly not a prebuilt rule. We take a similar approach with suppression where if the suppression fields are not included then suppression is not enabled for the rule. Do we need to explicitly mark non-prebuilt rules?

Reviewer 2: The data model has been redesigned so I'm resolving this thread.

    - When prebuilt rules are installed for the first time, set the `customized` field to `false`. Any update to the rule's cutomizable fields that causes it to diverge from the base version should, as a side effect, update the `customized` field to `true`.

```ts
// rule_schema/model/rule_schema.ts

[...]
export type BaseRuleParams = z.infer<typeof BaseRuleParams>;

> export const BaseRuleParams = z.object({

Reviewer: If we're doing migration on write, we need to differentiate between changes to the internal rule schema and the API schema. Migration on write means that the internal rule schema needs to represent that the new fields may not always exist, so `prebuilt` and `customized` must be optional. In the API schema, those fields can be required if we're comfortable with the breaking change. The conversion layer between internal and API schema will be responsible for taking rule objects where those fields may not exist and populating the correct value in the API response where they must exist.

This schema here appears to be the internal rule schema, which should not have `prebuilt` and `customized` listed as required if we're doing migration on write.

      - we can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed.
      - See implementation details below.
    - For the **duplicate** rule action:
      - since we will be creating a new rule on ES, we should create it with the new schema. 
      - Per definition: all duplicated rules will be `custom` rules. That means that for all rules -including prebuilt rules-, when duplicating, we need to set the new rule's `prebuilt`, `immutable` and `customized` value to `false`.
      - however, this action will not perform a migration-on-write of the original rule being duplicated for two reasons:
        - would violate the principle of least surprise for the endpoint
        - the current implementation of the endpoint does not modify the original rule. This gives us no window of opportunity to migrate the rule and save to ES without adding a considerable performance overhead.
      - See implementation details below.
    - All other type of actions should **not** perform migration-on-write:
        - Enable
        - Disable
        - Delete
        - Export: see Importing and Exporting Rules section below

> **NOTE:** Endpoints that depend on the Rules Client's methods need to modify the rules' params (prebuilt and customized) before the Rule Client methods call the`validateRuleTypeParams`, because that util validates the passed parameters using the validators defined in `x-pack/plugins/security_solution/server/lib/detection_engine/rule_types`. Since those are an intersection between `BaseRulesParams` and the specific rule type's fields, `prebuilt` and `customized` are non-optional, which would cause the util to reject the validation of the params for non-migrated rules.

Reviewer: `validateRuleTypeParams` is also called when rules run, which is another reason that the internal rule schema must not specify `prebuilt` and `customized` as required if we're doing migration on write. This will cause rules to fail at runtime.

Will need to update the `x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts` file, where the `convertRulesFilterToKQL` method is defined. This method is used both in the frontend and in the serverside, and translates rule filter options to KQL filters that Elasticsearch can understand.

Here, we need to update the KQL filters and the logic for fetching Elastic prebuilt and custom rules, relying on `rule_source_type` but with fallback to `immutable`:

_Source for `convertRulesFilterToKQL`: [x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts)_

In order to retrieve Elastic prebuilt rules, we can filter for rules in which `alert.attributes.params.rule_source_type` is `external`, with a fallback to rules in which `alert.attributes.params.immutable` is `true`, for rules which have not had their schema migrated and `rule_source_type` does not still exists for them.

> > In the future, we can further distinguish between `external` rules by their source origin. If we introduce a new field within `external_source` that defines their origin (such as `repo_name`, `repo_id`, or similar), we can further filter by Elastic prebuilt rules, or rules that are externally sourced from a repository handled by the user, etc.

Author: cc @banderror @marshallmain 

After some discussion with @banderror, we decided to introduce a top-level `rule_source_type` field which can take only the values `external` and `internal`. This way we can filter Prebuilt rules from custom rules without having to rely on a subfield of the optional `external_source` field.

For the scope of Prebuilt Rule Customization, knowing the value of `rule_source_type` will be enough to know if a rule is Elastic prebuilt or custom.

In the future, thinking about DaC, a rule might be `custom` but `external`ly sourced, so here I described how we can possibly handle that in the future. I think at that point we will be rethinking how we display our filters in the application anyways: probably not simply filtering prebuilt vs custom, but more of a multivalue filter by the rule's origin. But I these decisions will need product input and are more part of the DaC epic anyways.

Reviewer: We met on zoom to discuss this and I expressed my concern that `rule_source_type` introduces a relationship between sibling fields of the schema: `external_source` is only valid if `rule_source_type` is `external`, so we'd need to either enforce that constraint in additional checks everywhere we manage rules or encode that constraint in the schema by using `rule_source_type` as a discriminant. However, "additional checks everywhere" is bug prone because it's hard to ensure that this check is done in every place, and we already have one discriminant for rule params at the top level so adding another one is not easy.

Per our zoom discussion today, I'm now completely in favor of using the structure @banderror suggested with `type: 'internal' | 'external'` as a subfield of `ruleSource` and using `type` as the discriminant for the `ruleSource` object. It looks awkward at first with only `type` in `ruleSource` when `type` is `internal`, but I like how it makes the type very explicit and also makes it clear which rules have been migrated since every migrated rule should have `ruleSource`. This approach also avoids the problem with multiple discriminants for one object, since `type` is now a subfield of `ruleSource` so we have a top level `type` discriminant and a `ruleSource.type` discriminant that are independent.

I also briefly tested adding `ruleSource` to the API response schema and didn't see any problems with having a discriminated union for `ruleSource` within the broader rule discriminated union.

`x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.gen.ts`
```
export type ResponseFields = z.infer<typeof ResponseFields>;
export const ResponseFields = z.object({
  id: RuleObjectId,
  rule_id: RuleSignatureId,
  immutable: IsRuleImmutable,
  updated_at: z.string().datetime(),
  updated_by: z.string(),
  created_at: z.string().datetime(),
  created_by: z.string(),
  revision: z.number().int().min(0),
  related_integrations: RelatedIntegrationArray,
  required_fields: RequiredFieldArray,
  setup: SetupGuide,
  execution_summary: RuleExecutionSummary.optional(),
  rule_source: z.discriminatedUnion('type', [
    z.object({
      type: z.literal('external'),
      updated_at: z.string().datetime(),
    }),
    z.object({
      type: z.literal('internal')
    })
  ])
});
```

Secondly, check that the `security_detection_engine` Fleet package is installed and install it if it isn't. We will need the package to be installed to check if a rule that is being imported is an Elastic prebuilt rule.

Then, using the `rule_id` and `version`, attempt to fetch the corresponding `security-rule` asset from ES. 

**If a matching `rule_id` and `version` is found**, that means that the rule is an Elastic prebuilt rule, and we should therefore dynamically calculate the rule's `external_source` and `rule_source_type` fields.

- `rule_source_type`: should always be `'external'` if rule is Elastic prebuilt.
- `external_source`:
  - `source_updated_at`: can be retrieved from the corresponding `security-rule` asset.
  - `is_customized`: should be calculated based on the `security-rule` asset's field and the field rules that were part of the import request parameters. If any of them are different, i.e. have diverged from the base version, `is_customized` should be true.

Finally, using the import payload, plus the rule's `security-rule` asset fields and the calculated `rule_source_type` and `external_source` fields, create the rule, or update it if already exists in Kibana.

> **If a matching `rule_id` is found, but the `version` is not found**, reject the import for that rule. Without a version, we cannot dynamically calculate `external_source`.

Author: This is a case which we didn't discuss in the last Product meeting: what if a user sends as import parameters a `rule_id` that matches a `security-rule` asset, but the `version` doesn't match any historical (or the last) version?

Given that we've decided to "pay the cost" of making sure that the package is installed, I have tentatively decided here that the way to handle this edge case is to reject the import, clearly specifying why. Otherwise, we will fall in the pitfalls that @marshallmain explained [in this comment.](https://github.com/elastic/kibana/pull/171856#discussion_r1491794090)

**But** I'm not fully convinced this is the best UX, as a user won't be able to know which rule versions exist in Kibana, so not sure how they would attempt to "fix" this issue if they fell into this situation.

cc @banderror

Reviewer 2: @jpdjere Commented on that in https://github.com/elastic/kibana/pull/171856#discussion_r1524431230

**2. The user lowers the `version` field from the importing payload:** for example, a user exports a prebuilt rule with `version: 5`, and, before importing it, modifies it to `version: 4`. This wouldn't represent a problem as a new version of the rule (i.e. `version: 6`) would still trigger the update workflow, and the rule field diffs would still work as expected.

**3. The user increases the `version` field from the importing payload:** or example, a user exports a prebuilt rule with `version: 5`, and before importing it modifies it to `version: 6`. This would mean that when the actual rule with `version: 6` was released, we wouldn't trigger the update workflow, since the version from the target version (the update) needs to be higher than what is currently installed for that to happen. However, the update would eventually be triggered when the next update was released, i.e. when `version: 7` is released, and work as expected.

> **4. The user sets the `version` field from the importing payload to a number that does not exist in the Prebuilt Rules package**: this scenario would still depend on whether the `version` that the user modifies the rule to is higher or lower than the version of the update:

Author: Will need to update this scenario based on the discussion in [this comment](https://github.com/elastic/kibana/pull/171856#discussion_r1494542050)

_Status_: Finished Part 1, which includes sections from `Necessary rule schema changes` up until -and including- `Exporting and importing rules`.
 
 Covers:
  - rule schema changes
  - mappings
  - migration strategy and technical implementation
  - exporting and importing rules
  - schema-related changes needed in endpoints

_Pending_:

- Calculation of `isCustomized` field on endpoints that update/patch rules. Encapsulated reusable logic and how to implement it on those endpoints.
- Rework `Changes needed in the UI`
- Additional changes needed to `/upgrade/_review` and `/upgrade/_perform` endpoints
  - `/upgrade/_review`:
    - Concrete conflict solving algorithms for calculating merged versions
    - Update algorithms used for each field
  - `/upgrade/_perform`:
    - Change payload to allow user to select `merge` version of a rule, and send payload with customized values
- Three Way Diff Component:
  - UI
  - Frontend Logic

## Table of Contents

-   [Necessary rule schema changes](#necessary-rule-schema-changes)
    -   [`prebuilt` and `immutable` fields](#prebuilt-and-immutable-fields)
        -   [`isCustomized` subfield](#iscustomized-subfield)

>         -   [`elasticUpdateDate` subfield](#elasticupdatedate-subfield)

Author: Yes, with `pandoc`

-   [List of things to fix (create tickets)](#list-of-things-to-fix-create-tickets)
    -   [Rule fields](#rule-fields)
-   [Updating Prebuilt Rules](#updating-prebuilt-rules)
    -   [Changes to upgrade `_review` and `_perform` endpoints](#changes-to-upgrade-_review-and-_perform-endpoints)
    -   [Rule field diff algorithms](#rule-field-diff-algorithms)
-   [Changes in UI](#changes-in-ui)
-   [Scenarios for bulk accepting updates](#scenarios-for-bulk-accepting-updates)
-   [Marking rules (and their fields) as customized](#marking-rules-and-their-fields-as-customized)
-   [Other open questions](#other-open-questions)

## Necessary rule schema changes

In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing a new nested `prebuilt` field and deprecating the `immutable` field.

### `prebuilt` and `immutable` fields

The `prebuilt` field will be a top-level object field in our rule schema that will contain subfields with information relating to Elastic Prebuilt rules.

It will be an optional field, which will only be present for Elastic prebuilt rules.

> Its existence within a rule object will determine if a rule is a prebuilt rule (and its absence will determine that it is a custom rule).

Author: Yes, this is an issue. I described a workaround below but have very big doubts about it.

> Its existence within a rule object will determine if a rule is a prebuilt rule (and its absence will determine that it is a custom rule).

Reviewer: I think this is fine, but it could make more sense if we have a field inside `prebuilt` (or the renamed equivalent, see my other [comment](https://github.com/elastic/kibana/pull/171856/files#r1482151547) like a `repo_name` or `author` that is intuitively required and very commonly used as a filter ("give me all prebuilt rules from `Elastic`"). Then "give me all prebuilt rules" is equivalent to "give me all prebuilt rules from any repo" - `prebuilt.repo_name: *`.

In this first phase, the `prebuilt` will contain two subfields: `isCustomized` and `elasticUpdateDate`.

```ts
// PSEUDOCODE - see complete schema in detail below
{
  prebuilt?: {
    isCustomized: boolean;
    elasticUpdateDate?: Date;
  }
}
```

#### `isCustomized` subfield

The `isCustomized` field will be a boolean field that determines whether a Prebuilt Rule has been customized by the user, i.e. if any of its fields have been modified and diverged from the base version of the rule, which is the version that is installed from the Prebuilt Rules `security_detection_engine` Fleet package.

This means that the `isCustomized` subfield only makes sense and will be used for prebuilt rules, or rules whose `prebuilt` field exists. It is therefor a subfield of the `prebuilt` field.

For prebuilt rules, the `prebuilt.isCustomized` value will be initially set to `false` when a brand new rule is installed, but will be rewritten to `true` if a rule's field is edited and diverges from the value from the base version of the rule.

See section [Calculating the `isCustomized` flag]()

> #### `elasticUpdateDate` subfield

Reviewer: 100% agree with future proofing these fields. This field was one I was going to comment on as needing a rename to support externally managed rules from other sources, so I'm glad we're aligned there. 

I would also suggest renaming the `prebuilt` object to something more generic, maybe `external` or `external_source`? `prebuilt`, like `immutable`, is likely to become outdated and not match how we actually use the fields unless we call all rules managed through Detections as Code "prebuilt" rules.

We can add more subfields later, and I imagine those might be things like `external_source.repo_name: "elastic_prebuilt"` to differentiate Elastic Prebuilt rules from other externally managed rules that end up in `security-rule` SOs.

> #### `elasticUpdateDate` subfield

Reviewer 2: The data model has been redesigned and the fields have been renamed.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts)_

```ts
export const PrebuiltRuleAsset = BaseCreateProps.and(TypeSpecificCreateProps).and(
  z.object({
    rule_id: RuleSignatureId,
    version: RuleVersion,
    related_integrations: RelatedIntegrationArray.optional(),
    required_fields: RequiredFieldArray.optional(),
    setup: SetupGuide.optional(),
    elasticUpdateDate: ElasticUpdateDate.optional(), // new optional field
  })
);
```

### Deprecating the `immutable` field

In order to mark the `immutable` field as deprecated, and making sure that our application and API users are aware that the field has been deprecated and replaced by the `prebuilt` field, we will communicate this change in three ways:

1. via updates to the documentation of all endpoints that return `RuleResponse` types
2. via a deprecation warning in the OpenAPI schema, as detailed above
3. by adding a custom response header in said endpoints.

> The `immutable` field will afterwards be actually removed from our API endpoint responses and our application after a deprecation period that should give our users enough time to adapt to this change. The length of this deprecation period can vary, but around 4 months, or roughly two ESS releases, is a good starting point.

Author: 12 months is actually a better starting point. But even more makes sense.
- Track how many users actually use this field (How?)

- **Delete Rules - DELETE /rules**: transforms rule in response to `RuleResponse` type with the `internalRuleToAPIResponse` method.
- **Bulk Actions** - `POST /rules/_bulk_action`: all other action types not mentioned in the previous section will perform normalization-on-read before responding:
  - Enable
  - Disable
  - Delete
  - Export: see Importing and Exporting Rules section below
- **Export Rules** - `POST /rules/_export`:
  - this endpoint is unused via the UI, but might still be used via the public API
  - same logic as for `_bulk_action` with the `export` action explained in section [Exporting rules](#exporting-rules) below: only normalization-on-read will be performed before writing to the output `ndjson` file.

### Technical implementation of migration-on-write

The logic for the migration of the rule saved objects, which means the determination of the `immutable` and `prebuilt` fields before writing to ES, will be encapsulated in three helper methods:

- `migratePrebuiltSchemaOnRuleCreation` for rule creation
- `migratePrebuiltSchemaOnRuleUpdate` for rule updates (including patching)

> - `migratePrebuiltSchemaOnRuleBulkEdit` for bulk rule updates

Reviewer: Agree that the level of detail is too high. Focusing on **what** an API accomplishes from a user's perspective rather than **how** it does it would be helpful for the RFC content.

// so we need to search for the existence of the isCustomized subfield instead, which is required
export const KQL_FILTER_PREBUILT_RULES = `${PARAMS_PREBUILT_IS_CUSTOMIZED_FIELD}: *`;
export const KQL_FILTER_CUSTOM_RULES = `NOT ${PARAMS_PREBUILT_IS_CUSTOMIZED_FIELD}: *`;
export const KQL_FILTER_IMMUTABLE_RULES = `${PARAMS_IMMUTABLE_FIELD}: true`;
export const KQL_FILTER_MUTABLE_RULES = `${PARAMS_IMMUTABLE_FIELD}: false`;

interface RulesFilterOptions {
  filter: string;
  showCustomRules: boolean;
  showElasticRules: boolean;
  ...
}

export function convertRulesFilterToKQL({
  filter: searchTerm,
  showCustomRules,
  showElasticRules,
  ...
}: Partial<RulesFilterOptions>): string {
  const kql: string[] = [];

  // [... file continues ...]

  if (showCustomRules && showElasticRules) {
    // if both showCustomRules && showElasticRules selected we omit filter, as it includes all existing rules
  } else if (showElasticRules) {
    kql.push(`(${KQL_FILTER_PREBUILT_RULES} OR ${KQL_FILTER_IMMUTABLE_RULES})`);
  } else if (showCustomRules) {
    kql.push(`(${KQL_FILTER_CUSTOM_RULES} OR ${KQL_FILTER_MUTABLE_RULES})`);
  }

  // [... file continues below...]
```

> Ntocie that in order to retrieve rules in which the `alert.attributes.params.prebuilt` field exists, we must rely instead on the existence of `alert.attributes.params.prebuilt.isCustomized`. This is so because the `prebuilt` field lives inside the unmapped `params` field, and KQL syntax does not allow searching for the existence of an object like `prebuilt` with the recommended syntax `alert.attributes.params.prebuilt: *`.

Reviewer 2: The data model has been redesigned and this comment is no longer relevant.

### Rule Management endpoints

- **Create Rules** - `POST /rules` and **Bulk Create Rules** - `POST /rules/_bulk_create`:

Currently, we don't support the `immutable` field in any of the endpoints' request parameters (except for the Import endpoint). We shouldn't support the `prebuilt` field either, because this value should be controlled by the app on the server side, and not by users.

This is so because we will never want users to be able to create their own prebuilt rules, only install them, import them, and customize them. Also, a prebuilt rule should always be able to be compared to a `security-rule` asset distributed by Fleet, and receive updates from it, which would not be possible if a user creates its own prebuilt rules.

Specifically, these two endpoints should be able to create only custom rules which means their `params` will look like:

```ts
// [PSEUDOCODE]
{
  immutable: false,
  // prebuilt should not be in the created rule object,
  // since this endpoint should only create custom rules
}

> ```

Reviewer 2: Why users should even bother specifying any of that? The `immutable` and `prebuilt` fields should not be in the request schemas of the endpoints mentioned above:

**Create Rules** - `POST /rules` and **Bulk Create Rules** - `POST /rules/_bulk_create`

Author: Add a clarifiction that what is meant is the rule's SO parameters and not the request parameters of the endpoints

Both endpoints used the already discussed `createRules` method to create rules. The only changes needed for both are explicitly passing the new `isPrebuilt` argument as `false`:

> _Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/api/rules/create_rule/route.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/api/rules/create_rule/route.ts)_

Reviewer 2: Let's remove all these code examples for the sake of brevity.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_upgrade/review_rule_upgrade_route.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_upgrade/review_rule_upgrade_route.ts)_

```ts
// [... file continues above ...]

const calculateRuleInfos = (results: CalculateRuleDiffResult[]): RuleUpgradeInfoForReview[] => {
  return results.map((result) => {
    const { ruleDiff, ruleVersions } = result;
    const installedCurrentVersion = ruleVersions.input.current;
    const targetVersion = ruleVersions.input.target;

    // [... file continues ...]

    const targetRule: RuleResponse = {
      ...convertPrebuiltRuleAssetToRuleResponse(targetVersion),
      // Overwrite the prebuilt object returned from convertPrebuiltRuleAssetToRuleResponse
      // to account for the currently installed rule's value of `isCustomized`.
      prebuilt: {
        isCustomized: installedCurrentVersion.prebuilt?.isCustomized ?? false,
        elasticUpdateDate: targetVersion.elasticUpdateDate,

>       },

Reviewer 2: Why do we need to overwrite `prebuilt` here like that? I read the accompanying text, but didn't understand the reason.

Author: This doesn't make sense, we wouldn't even display this data to the user, they are techinal fields.

Remove.

```ts
// [... file continues above ...]

export const performRuleUpgradeRoute = (router: SecuritySolutionPluginRouter) => {
  router.versioned
    .post({
      access: 'internal',
      path: PERFORM_RULE_UPGRADE_URL,
      // [...]
    })
    .addVersion(
      {
        // [...]
      },
      async (context, request, response) => {
        // [...]

        try {
          // [...]

          // Endpoint will include logic for calculating which rule version should the current rule be upgraded to,
          // as well as calculating the isRuleCustomizedDuringUpgrade boolean.
          // Details explained further below in this document.

          // Perform the upgrade, but pass in `isRuleCustomizedDuringUpgrade` as an argument
          const { results: updatedRules, errors: installationErrors } = await upgradePrebuiltRules(
            rulesClient,
            targetRules,
            isRuleCustomizedDuringUpgrade

>           );

Reviewer 2: Shouldn't we calculate, if a rule has been customized during upgrade, per each rule separately? Why `isRuleCustomizedDuringUpgrade` is passed as a single value along with an array of `targetRules`?

    ) {
      return {
        statusCode: 200,
        rule: transformRuleToExportableFormat(internalRuleToAPIResponse(matchingRule)),
      };
    } else {
      return {
        statusCode: 404,
        missingRuleId: { rule_id: ruleId },
      };
    }
  });

// [... file continues below ...]
```

Notice that the **Bulk Actions** - `POST /rules/_bulk_action` does not perform a **dry run** in order to filter out prebuilt rules when the user runs an **export** action; this is the only check that is performed.

*Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts)*
```ts
// [... file continues above ...]
export const getExportAll = async (
  // [...]
): Promise<{
  rulesNdjson: string;
  // [...]
}> => {
  // const ruleAlertTypes = await getNonPackagedRules({ rulesClient }); <---- current code
  const allRules = await getRules({ rulesClient }); // <------ get all rules now
  const rules = transformAlertsToRules(allRules);

// [... file continues below ...]
```

> ### Importing rules

Reviewer: Some updates per today's discussion at the Simplified Protections WG:

@jpdjere made a great point that we need to allow users to specify the version on import in order for upgrades to work correctly. If a user exports a prebuilt rule from one cluster at version 3 and imports it to another cluster where an updated package has been installed with version 4 of that rule, and we import the rule without specifying the version, we won't be able to easily tell if the imported rule is version 4 customized or version 3 un-customized. If we import the rule and treat it as version 4 customized, but it should have been version 3 un-customized, then upgrading the rule to version 5 would require the user to merge "their customizations" that they didn't actually make. If we instead allow the rule to be imported as version 3 un-customized, the upgrade flow is easier as there are no customizations to worry about merge and we can simply fast forward to version 5 when they want to upgrade.

With that case in mind, I think the combination of `rule_id` + `version` would be the unique identifier we use on rule import to load and populate the additional `prebuilt` fields like updated at, `isCustomized`, etc based on the security-rule assets. With the addition of `version`, we can pick the right asset from the history instead of always using the latest one as I had imagined previously.

The core requirement that I had in mind when I proposed that the import API should not allow users to set the `prebuilt` fields was that the `prebuilt` fields must always be in sync with the associated `security-rule` asset - we need to do validation at the API boundaries so our application within those boundaries can always assume that the data in those fields is reliable. But maybe it would be convenient to allow users to include `prebuilt` fields in the ndjson they import to mark which rules should be linked to security-rule assets, and during the import process we could load the prebuilt rules package (if it is not already loaded) and either overwrite or reject rules in the import JSON with `prebuilt` fields that don't match up with what's in the package.

> ### Importing rules

Author: @marshallmain 

> But maybe it would be convenient to allow users to include prebuilt fields in the ndjson they import to mark which rules should be linked to security-rule assets

Not sure I understand why we'd need the data from the `prebuilt` field if we have the `rule_id` and the `version` and are fetching the rule asset based on those two values. If the `prebuilt` field can be completely recalculated from there, why would we allow `prebuilt` to be sent in the payload?

> ### Importing rules

Reviewer: If `prebuilt` can be passed in the payload then we don't need an extra parameter on the API, separate from the rule import data, to control whether or not we try to link imported rules to prebuilt rule assets - each rule's JSON tells us if it should be linked or not. None of the data like `isCustomized` or `sourceUpdatedAt` from the `prebuilt` object would be copied directly into the created rule, but the existence of the `prebuilt` object would signal the user's intent to import a rule as a prebuilt rule instead of a custom rule that just coincidentally shares a rule_id and version with a prebuilt rule.

With the `ruleSource` changes we talked about, if `ruleSource.type: 'external'` (or `ruleSource` is missing and `immutable: true`) then we'd fetch the rest of the `ruleSource` data from the security rule asset. If `ruleSource.type: 'internal` (or `ruleSource` is missing and `immutable: false`) then we wouldn't need to fetch any other data from the security rule assets. I think with `ruleSource.type` now the idea is a bit clearer, whereas before we would look for `prebuilt` but recalculate all the fields in it so the `prebuilt` data seems redundant (we would just be looking for the existence of the object), with `ruleSource` we keep `ruleSource.type` and recalculate the rest of the values if type is `external`.

2. **Calculation of the `is_customized` field**: this field which is part of the `rule_source` field for external rules needs to be recalculated whenever a rule's field is possibly modified. See implementation details in the [Updating the `is_customized` field](#updating-the-is_customized-field) section.
3. **Removing checks that block modifying prebuilt rules:** some of our endpoints prevent users from modifying prebuilt rules. This check needs to be removed from all endpoints to allow the customization of prebuilt rules.
4. **Blocking the update of non-customizable fields:** while the goal of this epic is to allow users to modify their prebuilt rule fields, there are certain rule fields that we will still want to prevent the user from modifying via endpoints, since we will not provide support for resolving conflicts for them via the UI or API if they arise during an update. See the **Customizable** column in [this ticket](https://github.com/elastic/kibana/issues/147239) for a detailed list of fields that should be blocked from modification via the endpoints.

The table below shows which of these changes need to be applied to our endpoints: (✅ change needed - ❌ no change needed)

>   <table>

Author: Replaced, thanks for the suggestion

| **Actions**    | - Actions<br>- Response actions (custom query rules only)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |

Out of these four tabs, only **Actions** is enabled and accessible when editing a prebuilt rule - since actions (and shared exceptions lists and snoozing) are the only fields that can currently be modified for prebuilt rules from the UI.

All of the fields in the UI, listed above, are currently editable for custom rules, except for rule type, which is read only.

Once done editing the rule, the user clicks on the "Save Changes" button, which calls the **Update Rule** - `PUT /rules` endpoint, passing the payload for the whole rule.

> **Expected behaviour for customizing prebuilt rules**

Author: This is not thought of for now; but I think it's a valid idea that we can validate with @approksiu to see if this is something that users would actually need.

If we implemented something like this, I see two possible approaches:
1. A new endpoint that fetches the asset and updates all the fields to the values provided there.
2. Reusing the `/install/_perform` endpoint, which would need some modifications to allow re-installing already installed rules, and keep all existing non-customizable fields.

> **Expected behaviour for customizing prebuilt rules**

Author: - Create ticket for this use case.

### Concrete field diff algorithms by type

The algorithm used in the `/upgrade/_review` endpoint for calculating rule version diffs' core structure is [fully implemented](https://github.com/elastic/kibana/pull/148392) and operational. However, the current method for calculating field differences is basic: it currently simply compares field values across versions and flags a conflict if the base version, the current version and the target versions differ from one another. In this conflict scenario, the `mergedVersion` of the field proposed by the algorithm always equals the `targetVersion`.

We propose developing more adaptable diff algorithms tailored to specific rule fields or rule field types. These more specific algorithms aim to minimize conflicts and automate merging of changes, doing the best effort to keep the intended customizations applied by the user, as well as the updates proposed by Elastic. Hopefully, the user will be able to simply review the proposal and accept it.

In general, for all types of fields, we will follow the heuristic:

> - if `base` === `target` && `base` !== `current`, we set merge proposal to be the current version, without a conflict

Author: Addressed

This means, however, that there will be a change in behaviour of the `immutable` field: with the release of the feature, rules with the `immutable: true` value will now be customizable by the user, which is not the current behaviour.

In order to mark the `immutable` field as deprecated, and making sure that our application and API users are aware that the field has been deprecated and replaced by the `rule_source` field, we will communicate this change in three ways:

1. via updates to the documentation of all endpoints that return `RuleResponse` types
2. via a deprecation warning in the OpenAPI schema, as detailed above
3. by adding a custom response header in said endpoints.

> The `immutable` field will afterwards be actually removed from our API endpoint responses and our application after a deprecation period that should give our users enough time to adapt to this change. The length of this deprecation period can vary depending on adoption or other factors, but around 18 monthscan be a good starting point.

Author: Makes sense. Addressed 👍

- [`createRules`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/create_rules.ts)
- [`patchRules`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/patch_rules.ts)
- [`updateRules`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/update_rules.ts)

This "overuse" of these 3 methods for a variety of user actions makes their logic tightly coupled and creates a considerable amount of complexity to CRUD functions that should remain logically simple.

For example: the `createRules` method is used in 3 use cases:
1. when creating custom rules with our Rule Creation endpoints
2. when importing rules, when the imported `rule_id` does not already exist in Kibana 
3. when upgrading rules, if a rule undergoes a `type` change, the existing rule is deleted a new one is created.

The same happens with `patchRules`. It is used:
1. when patching rules via our Rule Patch endpoints
2. when importing rules, when the imported `rule_id` is brand new

> 3. when upgrading rules, if a rule maintains its `type`.

Author: Ah, ok, I had forgotten about that updated. I removed that example here.

> I feel that we should be using updateRules when upgrading rules as well...

I'm not sure. I think we need to write new CRUD methods specific for each scenario. And not reuse a method like `updateRule`s for different user scenarios like: 1. editing a rule and 2. upgrading a rule.

Let's discuss this more in depth.

> 3. when upgrading rules, if a rule maintains its `type`.

Author: - Create ticket for replacing `PATCH` logic with `PUT` when upgrading rules.

| **Read Rule** - `GET /rules` | <center>✅</center> | <center>❌</center> |  |
| **Delete Rules** - `DELETE /rules` | <center>✅</center> | <center>❌</center> |  |
| **Export Rules** - `POST /rules/_export` | <center>✅</center> | <center>❌</center> | - Endpoints is unused in UI, still used via public API <br> - See section [Exporting rules](#exporting-rules) |
| **Import Rules** - `POST /rules/_import` | <center>✅</center> | <center>✅</center> | - See section [Importing rules](#importing-rules) |
| **Update Rule** - `PUT /rules`| <center>✅</center> | <center>✅</center> | - Used in the UI when updating/modifying a single rule via the Rule Editing page |
| **Patch Rule** - `PATCH /rules`| <center>✅</center> | <center>✅</center> | - Used in the UI for attaching shared exceptions list to rules |
| **Bulk Update Rules** - `PUT /rules/_bulk_update`| <center>✅</center> | <center>✅</center> | - Deprecated and unused by the UI. Might still be used by API users |
| **Bulk Patch Rules** - `PATCH /rules/_bulk_update`| <center>✅</center> | <center>✅</center> | - Deprecated and unused by the UI. Might still be used by API users |

> | **Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal) | <center>✅</center> | <center>✅</center> | - Current way of upgrading a prebuilt rule |

Author: That's right. Some of them should also perform migration-on-write as well. Adding them to the table.

|**Bulk Actions** - `POST /rules/_bulk_action`: | | | This endpoint includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`, since we never write to ES when this parameter is set to `true`.|
|  <li>_**Enable and disable action**_</li> | <center>✅</center> | <center>❌</center> | - Migration-on-write is technically possible but should be done on Alerting Framework side. |
| <li>_**Delete action**_</li> | <center>✅</center> | <center>❌</center> | - Deletes ES object but returns deleted rules data, so so normalization-on-read is enough. |
| <li>_**Export action**_</li> | <center>✅</center> | <center>❌</center> | - See section [Exporting rules](#exporting-rules) |
| <li>_**Duplicate rule action**_</li> | <center>❌</center> | <center>❌</center> | - Per definition, all duplicated rules will be `custom` rules. That means that all duplicated rules (the duplicates) are newly created and should get a `rule_source` of type `internal`, and no migration or normalization is neccessary. |

> | <li>_**Edit rule action**_</li> | <center>❌</center> | <center>✅</center> | - No normalization-on-read is needed since endpoint doesn't return whole rule object <br>- We can take advantage of the `ruleParamsModifier` to carry out the migration-on-write, regardless of the type of edit that is being performed. <br>- See implementation details in the [Bulk editing rules](#bulk-editing-rules) section.

Author: Corrected both points.

Updating rules can currently be performed via five endpoints:

- **Update Rule** - `PUT /rules`
- **Patch Rule** - `PATCH /rules`
- **Bulk Update Rules** - `PUT /rules/_bulk_update`
- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`
- **Bulk Actions** - `POST /rules/_bulk_action`: with **bulk edit** action

Upgrading prebuilt rules to their newer version is done by two endpoints:

- **(LEGACY) Install Prebuilt Rules And Timelines** - `PUT /rules/prepackaged`
- **Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal)

The legacy endpoint does not allow for customization of fields during the upgrade, but the new rule upgrade customization endpoint does. 

Additionally:

- **Bulk Actions** - `POST /rules/_bulk_action`: with **duplicate** action

will perform migration but does not allow for customization during the duplication process.

So we can analyze the expected outputs of the migration of all these 8 endpoints together.

The resulting values for `immutable` and `rule_source` when calling these endpoints, and the migration being performed in the background, should be as follows:

<table>
  <thead>
    <tr>
      <th>Migration use case</th>

>       <th>Current value of <pre>rule_source</pre></th>
>       <th>Current value of <pre>immutable</pre></th>
>       <th>Any field diverges <br> from base version after update?</th>
>       <th>Results</th>

Reviewer 2: Nit: In the "results" column we show a json containing the rule source and immutable parameters, but for the currently stored values we have two columns. I think we could improve readability of the table if we used a single column for the currently stored value.

```suggestion
      <th>Currently stored</th>
      <th>Any field diverges <br> from base version after update?</th>
      <th>Result to be stored</th>
```

Author: Fixed.

One note for when we discuss this: this examples consider that `immutable` is not deleted when the migration-on-write is done. As you mentioned in another comment, I think now it should actually be deleted. Let's discuss tomorrow this, but I can remove the `immutable` field where the rule was already migrated in this table.

    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Custom rule</b> (not migrated yet)</td>
      <td>
        <pre>undefined</pre>
      </td>
      <td><pre>false</pre></td>
      <td>N/A - Doesn't apply for custom rules</td>
      <td>
        <pre>
{
  rule_source: {
    type: 'internal'
  },
  immutable: false,
}
        </pre>
      </td>

>     </tr>

Author: Ah, just saw this comment. Replied in https://github.com/elastic/kibana/pull/171856#discussion_r1546807442

>     </tr>

Author: Addressed camelCase

}
        </pre>
      </td>
    </tr>
    <tr>
      <td><b>Prebuilt rule</b> (not yet migrated, with customizations)</td>
      <td>
        <pre>undefined</pre>
      </td>
      <td><pre>true</pre></td>
      <td>Yes</td>
      <td>
        <pre>
{
  rule_source: {
    type: 'external',
    isCustomized: true,
    ...
  },
  immutable: true,
}
        </pre>
      </td>
    </tr>
    <tr>
      <td><b>Prebuilt rule</b> (already migrated, no customizations)</td>
      <td>
        <pre> 
{
  type: 'external',
  isCustomized: false,
    ...
}
          </pre>
      </td>
      <td><pre>true</pre></td>
      <td>No</td>
      <td>
        <pre>
{
  rule_source: {
    type: 'external',
    isCustomized: false,
    ...
  },
  immutable: true,
}
        </pre>
      </td>
    </tr>
    <tr>
      <td><b>Prebuilt rule</b> (already migrated, with customizations)</td>
      <td>
        <pre> 
{
  type: 'external',
  isCustomized: true,
    ...
}
          </pre>
      </td>
      <td><pre>true</pre></td>
      <td>Yes</td>
      <td>
        <pre>
{
  rule_source: {
    type: 'external',
    isCustomized: true,
    ...
  },
  immutable: true,
}
        </pre>
      </td>

>     </tr>

Reviewer 2: The currently stored value of rule source should be this, I believe:

```js
{
  type: 'external',
  isCustomized: false,
    ...
}
```

Author: In this case, both examples work: if the rules is or  not customized and during the editing it gets customized, then isCustomized should be `true` in both cases.

But this makes me think if we should add one more case: if the initial value of `isCustomized` is `true` and the editing of the rule makes it match the `base` version, then the resulting `isCustomized` should be `false`. But maybe the detail is not needed in the RFC

Author: - Add scenario to table.

      <td>
        <pre> 
{
  type: 'external',
  isCustomized: true,
    ...
}
          </pre>
      </td>
      <td><pre>true</pre></td>
      <td>No</td>
      <td>
        <pre>
{
  rule_source: {
    type: 'external',
    isCustomized: false,
    ...
  },
  immutable: true,
}
        </pre>
      </td>
    </tr>
    <tr>
      <td>Invalid case: Migrating a migrated non-customized prebuilt rule, with customizations. <br><br> `immutable` should never be false if `rule_source.type' is 'external'. Migration should correct this inconsistency.</td>
      <td>
        <pre> 
{
  type: 'external',
  isCustomized: false,
    ...
}
          </pre>
      </td>
      <td><pre>false</pre></td>
      <td>Yes</td>
      <td>
        <pre>
{
  rule_source: {
    type: 'external',
    isCustomized: true,
    ...
  },
  immutable: true,
}
        </pre>
      </td>
    </tr>

  </tbody>
</table>

As explained before, all these endpoints suffer from the tightly-coupled logic problem described in the section [`Problem with tightly coupled logic in our endpoints`](#problem-with-tightly-coupled-logic-in-our-endpoints).

> We should therefore create new CRUD methods as needed to uncouple them logically and cleanly apply migrations to them.

Reviewer 2: Nit: I think there's no need to repeat the same information again.

    <td>
<pre>
{
  name: "My custom prebuilt rule",
} 
</pre>
    </td>
    <td><b>No</b></td>
    <td>
    Rule import rejected<br> because of rule_id match and overwrite flag is false
      </td>
    <tr>
  <tr>
    <td>
<pre>
{
  name: "My prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: false,
  }
} 
</pre>
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: true,
  }
} 
</pre>
      </td>
    <tr>
  <tr>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: true,
  }
} 
</pre>
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My prebuilt rule",
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: false,
  }
} 
</pre>
      </td>
    <tr>
  </tbody>
</table>

### Handling the `version` parameter

> In the code above, we made an additional check: if the rule to be imported is a prebuilt rule, then the `version` field will be required as part of the payload as well. Attempting to import a prebuilt rule without it will result in an error and the importing of that rule will be skipped. The `version` is necessary to be able to provide updates to the rule when a new version is available in the Prebuilt Rules package. Without it we don't know if the installed version is newer or older than the update being offered by the package.

Reviewer 2: Ok, so another important note. The `version` field should be required only for importing prebuilt rules. We should preserve the current behavior where users are able to import custom rules without specifying their versions.

In the current `RuleToImport` schema the `version` field comes from `BaseCreateProps` where it's defined as an optional field:

```
export const RuleToImport = BaseCreateProps.and(TypeSpecificCreateProps).and(
  ResponseFields.partial().extend({
    rule_id: RuleSignatureId,
    immutable: z.literal(false).default(false),
  })
);
```

I think this schema should stay the same with regards to the `version` field. We shouldn't override it and make the field required in `RuleToImport`. Otherwise, it would be a breaking change.

The rule import endpoint should:

- for custom rules being imported, if `version` is not specified, set it to 1
- for prebuilt rules being imported, if `version` is not specified, throw an error

Let's update this paragraph and the "Rule Import request schema" section accordingly.

Author: I updated the `#### Rule Import request schema` section to details this: 

- only `rule_id` is additionally required in the endpoint, `immutable` is not required anymore.
- whether the rule is prebuilt or not will be calculated based on the `rule_id` (searching in the `security-rule` SO for matches) (EDGE CASE TO RESOLVE: What if the package is not installed?)
- treatment of `version` in imported rules with the logic described above.

| **Actions**    | - Actions<br>- Response actions (custom query rules only)|

Out of these four tabs, only **Actions** is enabled and accessible when editing a prebuilt rule - since actions (and shared exceptions lists and snoozing) are the only fields that can currently be modified for prebuilt rules from the UI.

All of the fields in the UI, listed above, are currently editable for custom rules, except for rule type, which is read only.

Once done editing the rule, the user clicks on the "Save Changes" button, which calls the **Update Rule** - `PUT /rules` endpoint, passing the payload for the whole rule.

**Expected behaviour for customizing prebuilt rules**

- All four tabs of the Rule Edit page should be enabled, and all the fields within each tab should be editable, as they currently are for custom rules.
- The only fields in the UI that should not be customizable are: **Rule Type**, **Author** and **License**, which should continue to be read-only.
- **Definition** should be the default open tab when opening the edit rule page for a prebuilt rule (current default is **Actions**)
- Field validation should continue to work as it does for custom rules.

> - No fields should return a validation error for the values that come from the `security_detection_engine` package prebuilt rules. This means that a user should be able to successfully save the prebuilt rule with no changes. See **List of issues to fix** below.

Author: Applied

#### Via Rules Table page

> - The Rules Table's filter should be updated so that rules can be fitlered according to their status as **custom**, **Elastic prebuilt** and **customized Elastic prebuilt**. Notice that **customized Elastic prebuilt** will be a subset of **Elastic prebuilt**.

Author: I had this doubt as well, this got clarified in the last design review.

We propose developing more adaptable diff algorithms tailored to specific rule fields or rule field types. These more specific algorithms aim to minimize conflicts and automate merging of changes, doing the best effort to keep the intended customizations applied by the user, as well as the updates proposed by Elastic. Hopefully, the user will be able to simply review the proposal and accept it.

In general, for all types of fields, we will follow the heuristic:

- if `target` === `base` && `current` !== `base`, we set merge proposal to be the current version, without a conflict
- if `current` === `base` && `target` !== `base`, we set merge proposal to be the target version, without a conflict
- if `base` !== `current` !== `target`: mark the diff as a conflict AND:
  - **if conflict is non-solvable**: use the `current` version as the merge proposal 

>   - **if conflict is solvable** use the generated merged version as the merge proposal (possible only in a few types, detailed below - see tables)

Author: Updated the table.

      <td><pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "keyword"
  },
  {
    ecs: false,
    name: "host.os.ip",
    type: "ip"
  },
]    
      </pre></td>
      <td><pre>NO</pre></td>
    </tr>
    <tr>
      <td style="border-right:3px solid black">Customization and upstream update <b>solvable conflict</b> (ABC)<br><br> Object A modified in matching way</td>
      <td><pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "unknown"
  },
  {
    ecs: false,
    name: "host.os.type",
    type: "keyword"
  },
]      
      </pre></td>
      <td><pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "keyword"
  },
  {
    ecs: false,
    name: "new.field.name",
    type: "keyword"
  },
]    
      </pre></td>
      <td style="border-right:3px solid black"><pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "keyword"
  },
  {
    ecs: false,
    name: "new.field.ip",
    type: "ip"
  },
]   
      </pre></td>
      <td><pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "unknown"
  },
    {
    ecs: false,
    name: "new.field.name",
    type: "keyword"
  },
  {
    ecs: false,
    name: "new.field.ip",
    type: "ip"
  },

> ]

Author: Oops. Thanks for the catch!

Both epics have many areas of contact, and features/decisions in one need to take account features in the others.

This RFC includes some decision that will be taken in order to future-proof our archtecture for the upcoming DaC epic, but the vast majority will refer specifically to the Prebuilt Rules Customization epic. Elastic prebuilt rules will become a specific case of externally sourced rules: in the future, rules can be installed as Elastic prebuilt rules from our `detection-rules` repo, but also from any other repository handled by the user.

During this RFC, we will explicitly refer to case of Elastic prebuilt rules, but take into account that this will be only a specific case of all the possible external sources for rules.

## Necessary rule schema changes

In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing the new top level field: the nested `rule_source` field, as well deprecating the `immutable` field.

```ts
// PSEUDOCODE - see complete schema in detail below
{
  rule_source: {
    type: 'external'

>     is_customized: boolean;

Author: As we discussed in the last meeting, given the pros and cons of using single world names for props in the long term, we'll use propoerty names that include underscores if needed.

In the case of `is_customized` we feel that it better indicates that the property is a boolean than simply `customized`, which can additionally be confused with other proerties that could come up in the future, like a `customizations` array.

    source_updated_at?: Date;
  } | {
    type: 'internal'
  },
}
```

### `rule_source` field

The `rule_source` field will be a required top-level object field in our rule schema that will be a discriminated union of two types: `'internal'` and `'external'`.

Rules with `rule_source` of type `internal` are rules generated internally in the Kibana application and which don't have an external origin outside of it.

Rules with `rule_source` of type `external` are rules who have an external origin or source. Elastic prebuilt rules are a case of `external` rule, with the `detection-rules` repository handled by the TRaDE team being their external source origin.

> This also means that a rule with this type of `rule_source` will determine that the rule is an Elastic Prebuilt Rule that can receive upstream updates via Fleet. This field is intented to partly replace the currently existing `immutable` field, which is used today for the same purpose, but -as its name indicates- also currently determines if a rule's fields can be modified/customized.

Reviewer 3: misprint `intented` -> `intended`

As seen in the schema, when the `rule_source` has type `'external'`, the object can contain two other subfields as well, `is_customized` and `source_updated_at`, explained below. If the object has type `'internal'`, no other fields should be present (subject to change in the future).

### Deprecating the `immutable` field

In the current application's state, rules with `immutable: false` are rules which are not Elastic Prebuilt Rules, i.e. custom rules, and can be modified. Meanwhile, `immutable: true` rules are Elastic Prebuilt Rules, created by the TRaDE team, distributed via the `security_detection_engine` Fleet package, and cannot be modified once installed.

> When successfully implemented, the `rule_source` field should replace the `immutable` field as a mechanism to mark Elastic prebuilt rules, but with one difference: the `rule_source` field will determine if the rule is an Elastic Prebuilt Rule or not, but now all rules will be customizable by the user in Kibana, i.e. independently of the existence (or absence) of `rule_source`.

Reviewer 3: Would it be more generic to use term `external rules` instead of Elastic Prebuilt Rules? Since we want to take into account Detection-as-Code general terms will help to keep the document relevant and avoid confusion as long as possible.

Author: There's a note at the begginning of the RFC that explains this: we want to keep the discussion centered around prebuilt rules, although the changes that we are making are as forward-compatible as possible with DoC.

Reviewer 3: @jpdjere sorry, I didn't get if your answer is yes or no. If it's yes to simplify the document Elastic Prebuilt Rules can be omitted everywhere besides the document beginning and replaced with external rules.

Author: The answer is no. We want to refer to Elastic prebuilt rules for now as this is what the epic relates to. When we start dealing with DaC, we can generalize the term to "external rules".

Because of this difference in the behaviour of the `rule_source` and `immutable` fields, we feel that a field called `immutable` will lose its meaning over time and become confusing, especially for consumers of the API who interact directly with the field name. That's why we want to eventually deprecate it and fully replace it with `rule_source`.

To ensure backward compatibility and avoid breaking changes, we will deprecate the `immutable` field but keep it within the rule schema, asking our users to stop relying on this field in Detection API responses. During the migration period, we want to keep the value of the `immutable` field in sync with the `rule_source` fields: this means that for all rules that have a `immutable` value of `true`, the `rule_source` field will always be of type `'external'`. Viceversa, rules with `immutable: false` will have a `rule_source` field of type `'internal'`.

This means, however, that there will be a change in behaviour of the `immutable` field: with the release of the feature, rules with the `immutable: true` value will now be customizable by the user, which is not the current behaviour.

> ### Subfields in the `rule_source` field

Reviewer 3: nit: Have you considered organizing fields in a table for better visibility?

> In this first phase, the `rule_source` will conditionally contain three subfields: the desciminant `type`, and the `is_customized` and `source_updated_at` values, only present when `type` is `'external'`.

Reviewer 3: typo `desciminant` -> `discriminant`

#### `type` subfield

> The `type` subfield serves as the discriminat to the disciminated union field `rule_source` and can take two values, as explained above: `'internal'` and `'external'`.

Reviewer 3: typo `discriminat` -> `discriminant`, `disciminated` -> `discriminated`

See the detailed explanation for this mechanism in the [Exporting and importing rules](#exporting-and-importing-rules) sections.

The `immutable` and `rule_source` fields will be ignored if passed in the request payload.

_Source: [x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts)_

```ts
// [... file continues above...]
import {
  // [...]
  RuleSignatureId,
  IsRuleImmutable,
  ExternalSourceAttributes,
  RuleVersion,
} from '../../model/rule_schema';

export type RuleToImport = z.infer<typeof RuleToImport>;
export type RuleToImportInput = z.input<typeof RuleToImport>;
export const RuleToImport = BaseCreateProps.and(TypeSpecificCreateProps).and(
  ResponseFields.partial().extend({
    rule_id: RuleSignatureId,
    version: RuleVersion,
  })
);
```

#### Internal rule schema

> **The internal rule schema** needs to represent that the new `ruleSource` field may not always exist, so both need to be optional.

Reviewer 3: What does `both` mean in `so both need to be optional`?

Reviewer 2: ++

Author: This is outdated from when we had more than one new property. Thanks for the catch.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts)_

```ts
export type BaseRuleParams = z.infer<typeof BaseRuleParams>;
export const BaseRuleParams = z.object({
  // [...]

  immutable: IsRuleImmutable,
  ruleSource: RuleSource.transform(camelize).optional(),
  // [...]
});
```

> > Notice that in the internal schema we cannot simply reuse the `RuleSource` attribute defined for the API schema, since it should have CamelCase and the API schema uses snake_case. We need to apply a transformation to our Zod type. See this [issue](https://github.com/colinhacks/zod/issues/486#issuecomment-1567296747).

Author: Refer to https://github.com/elastic/kibana/pull/171856/files?diff=unified&w=0#r1528319597

##### Prebuilt Rule asset schema

The `PrebuiltRuleAsset` type needs to be updated to include the new `source_updated_at` date that will be progressively shipped with new versions of rules in the Elastic Prebuilt Rules package.

Notice that this field will be a **top-level field in the Prebuilt Rule asset** schema, but will be part of the `rule_source` field in the API rule schema and internal rule schema.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts)_

```ts
export const PrebuiltRuleAsset = BaseCreateProps.and(TypeSpecificCreateProps).and(
  z.object({
    rule_id: RuleSignatureId,
    version: RuleVersion,
    related_integrations: RelatedIntegrationArray.optional(),
    required_fields: RequiredFieldArray.optional(),
    setup: SetupGuide.optional(),
    source_updated_at: SourceUpdatedAt.optional(), // new optional field
  })
);
```

### Deprecating the `immutable` field

> In order to mark the `immutable` field as deprecated, and making sure that our application and API users are aware that the field has been deprecated and replaced by the `rule_source` field, we will communicate this change in three ways:

Author: No. As we discussed in the last Tech Time meeting, since we are not introducing any breaking changes to the API we won't be using API versioning. This avoids the complexity of maintaing two different versions, with the new one being backwards compatible with the older one.

We propose developing more adaptable diff algorithms tailored to specific rule fields or rule field types. These more specific algorithms aim to minimize conflicts and automate merging of changes, doing the best effort to keep the intended customizations applied by the user, as well as the updates proposed by Elastic. Hopefully, the user will be able to simply review the proposal and accept it.

#### Concrete field diff algorithms by type

Depending on the specific field or type of field we might want to apply a specific merging algorithm when conflicts arise. Let's propose different types.

##### String fields

For string fields, in case of three-way conflicts, we will make a best effort to reconcile any modifications done by the user and any updates proposed by Elastic.

Examples: `name`, `description`, `setup`, `note` (Investigation guide)

  <table>
    <thead>
      <tr>
        <th style="border-right:3px solid black">Use case</th>
        <th>Base version</th>
        <th>Current version</th>
        <th style="border-right:3px solid black">Target version</th>
        <th>Merged version (output)</th>
      </tr>

>     </thead>

Reviewer 2: Let's have two tables: for single-line and multi-line string fields.

Please also add one more column to both tables: a boolean value indicating whether the result of the merge is a conflict or not. Please also add this column to other tables below.

Author: Added

        <td><code>[Security Solution] Prebuilt rule name</code></td>
        <td style="border-right:3px solid black"><code>Updated prebuilt rule name</code></td>
        <td><code>[Security Solution] Updated prebuilt rule name</code></td>
      </tr>
      <tr>
        <td style="border-right:3px solid black">Keep user-added suffixes and accept other modifications</td>
        <td><code>Prebuilt rule name</code></td>
        <td><code>Prebuilt rule name - (Custom version 3)</code></td>
        <td style="border-right:3px solid black"><code>Updated prebuilt rule name</code></td>
        <td><code>Updated prebuilt rule name - (Custom version 3)</code></td>
      </tr>
      <tr>
        <td style="border-right:3px solid black">Keep user customizations on conflicts, but add updates in sections with no conflicts</td>
        <td><code>Prebuilt rule name</code></td>
        <td><code>Customized prebuilt rule name</code></td>
        <td style="border-right:3px solid black"><code>Updated prebuilt rule name. Much better.</code></td>
        <td><code>Customized prebuilt rule name. Much better.</code></td>
      </tr>
    </tbody>
  </table>

##### Number fields

> Number fields should be treated as a whole unit, i.e. not breakable by digits. Therefore, there is only one possibility of conflicts, the scenario (A B C). In that case, **keep the current version** with the user customization.

Reviewer 2: Are we going to make the "if conflict use the current version as the merged version" approach universal to all rule fields and diff algorithms? If so, let's explicitly propose this in the RFC at the beginning of the "Changes to upgrade _review endpoint" section.

Author: Explanation added to the text intoduction of section `### Concrete field diff algorithms by type`

Examples: `risk_score`, `max_signals`

  <table>
    <thead>
      <tr>
        <th style="border-right:3px solid black">Use case</th>
        <th>Base version</th>
        <th>Current version</th>
        <th style="border-right:3px solid black">Target version</th>
        <th>Merged version (output)</th>
      </tr>
    </thead>
    <tbody align="center">
      <tr>
        <td style="border-right:3px solid black">Keep customizations if conflict</td>
        <td><code>50</code></td>
        <td><code>65</code></td>
        <td style="border-right:3px solid black"><code>20</code></td>
        <td><code>65</code></td>
      </tr>
    </tbody>
  </table>

##### Array fields

> ###### Array of strings fields

Reviewer 2: I think this is too much nesting which makes it hard to navigate the RFC. Look at the TOC preview:

<img width="636" alt="Screenshot 2024-03-18 at 15 02 02" src="https://github.com/elastic/kibana/assets/7359339/76effd93-59d3-47b6-89fa-268f0c2db263">

The following structure would be much simpler and easier to follow:

- Upgrading prebuilt rules
  - Changes to `upgrade/_review` endpoint
  - Changes to `upgrade/_perform` endpoint
  - Changes to diff algorithms
    - Strings
    - Numbers
    - Arrays of strings
    - Arrays of objects
    - MITRE ATT&CK mappings

Author: Simplified

For **array of strings** fields, conflict resolution will take place on an element by element basis - that means that the algorithm will not consider or try to solve changes within a single element. For example:

```
base: [test1]
current: [my-test1]
target: [test2]

> output: [my-test1]

Reviewer 2: I don't understand the logic behind this example. Can you elaborate what is it and how that algorithm should work (at a very high level)?

My expectation would be the following:

```
base: [linux, network]
current: [windows, host]
target: [linux, ml]
output: [windows, host, ml]
```

The logic is a bunch of simple set operations:

- Changes to `current` compared against `base`:
  - Added: [`windows`, `host`]
  - Removed: [`linux`, `network`]
- Changes to `target` compared against `base`:
  - Added: [`ml`]
  - Removed: [`network`]
- In combination:
  - Added: [`windows`, `host`, `ml`]
  - Removed: [`linux`, `network`]
- Which, after applying to `base` ([`linux`, `network`]), gives the result [`windows`, `host`, `ml`]

Author: Fixed this whole section.

@banderror @maximpn 
I do have one disagreement with this logic though, and would like to ask you for your opinion.

In the example given above, the user removed [linux, network] and added [windows, host], while the Elastic updated added [ml] and removed [network].

I agree that the end result should include all the added elements in both modifications so:
[windows, host] + [ml] -> [windows, host, ml] should be added.

But I think that the logic for removals should probably be different. I see two possibilities:
1. Remove the element if it was removed in BOTH operations. This would mean:
[linux, network] + [network] -> only [network] should be removed (and linux shouldn't). So basically, an **intersection** of both removal operations.

2. Remove the element if it was removed in the current version. The UX logic behind this is that we probably don't want to remove an index patter (for example) if the user has not expressly deleted it from their rule, just because the Elastic update proposes it. That might break the rule. In this case, it would work so:
```js
const base = new Set(["linux", "network"]);  
const current = new Set(["linux", "ml", "security]);  
const target = new Set(["aws", "cloudfront"]);  

const addedCurrent = current.difference(base); // [ml, security]
const removedCurrent = base.difference(current); // [network]

const addedTarget = target.difference(base); // [aws, cloudfront]

const merged = base
                .difference(removedCurrent)
                .union(addedCurrent)
                .union(addedTarget)

// Results in: [linux, ml, security, aws, cloudfront]
```
I think this is a more conservative approach, less-riskier for users and less prone to accidental broken rules.

What do you think? The RFC currently has option 1, but I think I'd rather change the approach to option 2.

Reviewer 3: @jpdjere I'm afraid code example you provided in the second case isn't fully reflecting the proposal. Both `current` and `target` versions don't have `network` but you say `Remove the element if it was removed in the current version. Or I just didn't what you meant here.

Reading explanations we have two options

1. remove an element if it was removed in `current` and `target` versions
2. remove an element if it was removed only in the `current` version

I'd say if user removed something it shouldn't appear again. So the second case makes sense.

Reviewer 2: @jpdjere @maximpn Here are my thoughts on that:

> Remove the element if it was removed in BOTH operations.
> remove an element if it was removed in `current` and `target` versions

Why this would be needed? If I, as a user, decided to remove a certain tag from an array of tags (maybe I decided to completely replace the prebuilt tagging system with my own custom set of tags), and the auto-merge algorithm adds it back for me just because Elastic didn't remove this tag, wouldn't this annoy me?

From a different angle: why my user customizations (deletions, in this case) MUST be echoed by updates from Elastic in order to be taken into account during rule upgrade?

From another different angle: if Elastic removed a tag (because, for example, Elastic decided to rework the prebuilt tagging system), why this change MUST be echoed by the same user customization in order to be taken into account during rule upgrade?

> Remove the element if it was removed in the current version.
> remove an element if it was removed only in the `current` version
> The UX logic behind this is that we probably don't want to remove an index pattern (for example) if the user has not expressly deleted it from their rule, just because the Elastic update proposes it. That might break the rule.

Why do you think updates from Elastic should be "weaker" than user customizations, when it comes to removing items from arrays? What if as part of fixing or tuning a rule Elastic needs to remove a certain index pattern, if it was previously added by mistake? Or replace a broader index pattern by a narrower index pattern (which would be == a removal + an addition)? Or, as mentioned above, what if Elastic wants to rework the set of prebuilt tags, and replace most of existing tags with new ones? Prebuilt tags were completely reworked a few months ago btw.

> const base = new Set(["linux", "network"]);  
> const current = new Set(["linux", "ml", "security"]);  
> const target = new Set(["aws", "cloudfront"]);  
> Results in: [linux, ml, security, aws, cloudfront]

Isn't it the same as `current.union(target)`? Not sure why the proposed code is more complicated than that.

----------

Considering that I'd stick to my proposal.

@approksiu Can you please review this thread of comments and share your opinion? This is something that can have an impact on UX.

Reviewer 2: From yet another angle:

> remove an element if it was removed in BOTH current and target versions
> remove an element if it was removed only in the current version

Just an example, what would happen if the diff algo would use either if these two heuristics:

```
base: [logs-*]
current: [logs-*]
target: [logs-endpoint.events-*]
output: [logs-*, logs-endpoint.events-*]
```

Author: - Revert back to first algorithm, but update Conflict to `SOLVABLE`

Therefore, we should prefer to keep the user's version for the whole block.(See this example in the first element of the array in the first row of the table below).

Examples: `related_integrations,` `required_fields`

###### Proposed algorithm

- Do element by element comparison:
  - If element index exists in the three versions and the values:
    - do not differ between all versions (base, current and target), **keep current** version. (A A A)
    - do not differ between base and current, but updates target, **keep target** version. (A A B)
    - differ between base and current, but doesn't update target, **keep current** version. (A B A)
    - differ between base and current, updates target, **keep current** version. (A B B)
    - differ between all version (base, current and target), **keep current** version. (A B C)
  - If element index exists in base and current version, but not in target:
    - If base differs from current: **keep current** version. (A B _)
    - If base equals current: **remove element**.  (A A _)
  - If element index exists in base and target version, but not in current:
    - **Remove element**. (A _ B) and (A _ A)

>   - If element index exists in current and target version, but not in base:

Reviewer 2: We shouldn't be comparing elements in the base, current and target arrays index-by-index. At least in most cases. In most cases, the order of elements doesn't matter: the order of tags in the tags array, the order of required fields, the order of related integrations, etc.

For example, these arrays of required fields should look completely equivalent to the diff algorithm:

```
base: [
  { ecs: true, name: 'host.name', type: 'keyword' },
  { ecs: true, name: 'host.ip', type: 'ip' },
  { ecs: false, name: 'custom_field', type: 'text' }
]
current: [
  { ecs: true, name: 'host.ip', type: 'ip' },
  { ecs: true, name: 'host.name', type: 'keyword' },
  { ecs: false, name: 'custom_field', type: 'text' }
]
target: [
  { ecs: true, name: 'host.name', type: 'keyword' },
  { ecs: false, name: 'custom_field', type: 'text' },
  { ecs: true, name: 'host.ip', type: 'ip' }
]
```

Which probably means the following:

- When order doesn't matter, we should be either a) using sets or maps instead of arrays under the hood when calculating diffs, or b) sorting arrays before comparing them.
- For some complex object values that don't have an explicit id, we should define an object's key for the diff algo. For example:
  - for required fields the key could be the name of the field
  - for related integrations the key could be the package name + the optional integration name
- For complex obects that already have a clear explicit id, we should be using this id as the object's key.
- The algorithm should be tracking changes to elements in the arrays by their keys.

The algorithm could have a similar structure to the one for arrays of scalar values (strings, numbers, etc):

- Given base is [element with key `A`, element with key `B`]
- Calculate changes to `current` compared against `base`:
  - Added: [element with key `C`]
  - Removed: [element with key `A`]
  - Modified: [element with key `B`]
- Changes to `target` compared against `base`:
  - Added: [element with key `D`]
  - Removed: [element with key `A`]
  - Modified: [element with key `B`]
- In combination:
  - Added: [element with key `C`, element with key `D`] (no conflicts)
  - Removed: [element with key `A`] (no conflicts)
  - Modified: [element with key `B`, element with key `B`] (conflict if these 2 are not exactly the same modification)

Author: Reworked whole section using this logic.

I have the same question for the removal of elements that I wrote in https://github.com/elastic/kibana/pull/171856#discussion_r1537371632

    - If base differs from current: **keep both current and target**. (_ A B)
  - If element index exists only in base version:
    - **Remove element** (A _ _)
  - If element index exists only in current version:
    - **Keep current version** (_ A _)
  - If element index exists only in target version:
    - **Keep target version** (_ _ A)
    
<table>
    <thead>
      <tr>
        <th style="border-right:3px solid black">Use case</th>
        <th>Base version</th>
        <th>Current version</th>
        <th style="border-right:3px solid black">Target version</th>
        <th>Merged version (output)</th>
      </tr>
    </thead>
  <tbody valign="top">
    <tr>
      <td rowspan=2 style="border-right:3px solid black">Keep user changes if element conflicts, but apply changes in elements with clean updates <br><br> <ul><li><b>1st elem: (A B C - keep current) </li><li>2nd elem: (A A B - keep target)</b></li></ul><br><br> Same for: <br><ul><li><b>1st elem: (A B C) + 2nd elem: (A B A - keep current)</b></li><li><b>1st elem: (A B C) + 2nd elem: (A B B - keep current/target)</b></li></ul></td>
      <td>
      <pre>
[
  logs-*,
  logstash-*

> ]

Reviewer 2: Why do we describe examples of changes to index patterns in the "Array of objects fields" section and not in the "Array of strings fields"? Please move it to "Array of strings fields" to a separate table.

Reviewer 2: Can't agree with the proposed algorithm for index patterns either:

<img width="1394" alt="Screenshot 2024-03-18 at 15 12 32" src="https://github.com/elastic/kibana/assets/7359339/664ac763-91d9-4769-ad77-2a08639e8724">

In the example above I'd expect the result to be:

```ts
['cluster_one:logs-*', 'cluster_one:logstash-*', 'new-index1-*', 'new-index2-*']
```

The logic would be simple: https://github.com/elastic/kibana/pull/171856#discussion_r1528519565

| **Actions**    | - Actions<br>- Response actions (custom query rules only)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |

Out of these four tabs, only **Actions** is enabled and accesible when editing a prebuilt rule - since actions (and shared exceptions lists and snoozing) are the only fields that can currently be modified for prebuilt rules from the UI.

All of the fields in the UI, listed above, are currently editable for custom rules, except for rule type, which is read only.

Once done editing the rule, the user clicks on the "Save Changes" button, which calls the **Update Rule** - `PUT /rules` endpoint, passing the payload for the whole rule.

**Expected behaviour for customizing prebuilt rules**

> - All four tabs of the Rule Edit page should be enabled, and all the fields within each tab should be editable, as they currently are for custom rules.

Reviewer 2: On the Rule Editing page, are we going to have any indicators that the rule is prebuilt and (if this is the case) customized? If the rule is customized, are we going to show what fields have been customized and how? The latter could be achieved via showing a separate tab with a diff between the current and the base versions.

Author: Per the discussion we had in the last Design Review, the first iteration will only show in the Rules Detail page the "status" of the rule (custom, Elastic prebuilt or customized Elastic prebuilt). I have added this description below in the "Via Rules Detail page" section. Initially we won't show which fields were customized and how.

Having said that, I think this is a feature that eventually will be needed/requested. Howeverm since it will required the design of a new endpoint to calculate the diffs between the current version and the base version, I don't want to continue adding to this RFC. I think we can do a follow-up RFC or even a POC for it once Milestone 3 is done or nearing completion.

What do you think?

Reviewer 2: @jpdjere That sounds good to me 👍

- Select a **timeline template** for the rule

As explained above, the UI validates that the five actions listed above are only possible to custom rules by using the `dryRun` mode of the **bulk actions** endpoint. If any of the selected rules in the request payload is a prebuilt rule, a message is displayed to the user informing them that modifying prebuilt rules is not possible, and that rule is removed from the subsequent request that is made in normal mode to actually edit the rule.

**Expected behaviour for customizing prebuilt rules**

- The Bulk Actions UI should now allow users to edit prebuilt rules.
- All bulk actions actions listed above should now be applicable to prebuilt rules (This means that the changes in the API for the validation done in `dryRun` mode should not remove the rules from the payload anymore).

> The `dryRun` mode request will still be necessary since we perform other checks that can still prevent prebuilt rules from being edited. For example, Machine Learning and ES|QL rules cannot be bulk edited to add or delete index patterns, since they don't have that field in their schema.

#### Via the Rules Details Page

> The Rules Details page allows the user to modify the rule in two ways:

Reviewer 2: Can we also describe in the RFC what changes to the UI we will need to introduce for the Rule Details and Rule Management pages?

Author: Added

- enable/disabling
  - uses `POST /rules/_bulk_action` endpoint with enable/disable action
- setting rule snoozing
  - uses `POST /alerting/rule/{rule_id}/_snooze` endpoint

Both of these actions are already possible for prebuilt rules, **so no changes are necessary here.**

#### Via the Shared Exception Lists page

The Shared Exception Lists page allows a user to create shared exception lists and then attach them to (or remove them from) a user's security rules.

This page calls the **Patch Rule** - `PATCH /rules` endpoint to modify only the `exceptions_list` field of a rule saved object. Since a user can append an exception list to multiple rules via the UI, the endpoint is called multiple times simultaneously, one for each rule that is attached to or removed from.

There is no validation regarding prebuilt rules in the UI for this use case: both the endpoint and the application currently allow the user to attach exceptions lists to both custom and prebuilt rules.

> This means that **no changes will be needed in this page.**

Reviewer 2: Just want to double-check that we're on the same page regarding this: adding / editing / removing exceptions should not have any impact on `is_customized`. If `is_customized === true` or `is_customized === false`, it should stay `true` or `false` (correspondingly) in the following cases:

- a new shared exception list is added to the rule
- a shared exception list is removed from the rule
- an exception item is added to the rule's shared exception list
- a default exception list is created for the rule and an exception is added to it
- an exception item is removed from the rule's default exception list
- etc

Please document this expected behavior in the RFC.

Author: Yes, we're on the same page here. I'm detailing this behaviour in the `Calculating the is_customized field` section.

      <th>Output</th>
    </tr>
  </thead>
  <tbody>

  
  <tr>
    <td>Not installed</td>
    <td>No</td>
    <td>N/A</td>
    <td>
<pre>
{
  name: "My custom rule",
} 
</pre>
    </td>
    <td>N/A</td>
    <td>
<pre>
{
  name: "My custom rule",
  rule_source: {
    type: "internal"
  }
} 
</pre>
      </td>
    </tr>
  

  <tr>
    <td>
<pre>
{
  name: "My custom rule",
} 
</pre>
    </td>
    <td>No</td>
    <td>N/A</td>
    <td>
<pre>
{
  name: "My edited rule",
} 
</pre>
    </td>
    <td>
      <b>No</b>
    </td>
    <td>
      Rule import rejected<br> because of rule_id match
    </td>
    <tr>

  <tr>
    <td>
<pre>
{
  name: "My custom rule",
} 
</pre>
    </td>
    <td>No</td>
    <td>N/A</td>
    <td>
<pre>
{
  name: "My edited rule",
} 
</pre>
    </td>
    <td>
      Yes
    </td>
    <td>
<pre>
{
  name: "My edited rule",
  rule_source: {
    type: "internal"
  }
} 
</pre>
      </td>
    <tr>

  <tr>
    <td>
      Not installed
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  source_updated_at: "2024-05-..."
} 
</pre>
    </td>
    <td><b>No</b></td>
    <td>
<pre>
{
  name: "My prebuilt rule",
} 
</pre>
    </td>
    <td>N/A</td>
    <td>

>       Prebuilt rule import rejected because no matching <br>

Reviewer 4: I am not sure about this scenario - it seems to be something discussed in https://github.com/elastic/kibana/pull/171856#discussion_r1524431230, but it seems not to be updated in the table here.

Author: Thanks, I missed updating this row. Updated now

      <td>
      <pre>
[
  logs-*,
  logstash-*
]
        </pre>
      </td>
      <td>
        <pre>
[
  cluster_one:logs-*,
  cluster_one:logstash-*
]
        </pre>
      </td>
      <td style="border-right:3px solid black">
      <pre>
[
  new-index1-*,
  new-index2-*,
]
        </pre>
        </td>
      <td>
      <pre>
[
  cluster_one:logs-*,
  cluster_one:logstash-*
]
        </pre>
      </td>
    </tr>
    <tr>
      <td>
      <pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "unknown"
  },
  {
    ecs: false,
    name: "host.os.type",
    type: "keyword"
  },
]
        </pre>
      </td>
      <td>
        <pre>
[
  {
    ecs: true,
    name: "event.UPDATED",
    type: "unknown"
  },
  {
    ecs: false,
    name: "host.os.type",
    type: "keyword"
  },
]
        </pre>
      </td>
      <td style="border-right:3px solid black">
      <pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "keyword"

>   },

Reviewer 2: I think this situation should end up with a conflict, because:

- In the current version, `event.action` was replaced by `event.UPDATED`, in other words `event.action` was removed and `event.UPDATED` was added.
- In the target version, `event.action` was updated: its `type` was changed from `unknown` to `keyword`.

It shouldn't be possible to simultaneously remove and update the same element.

What should be in the merged version in this case? Probably just the current version.

  {
    ecs: false,
    name: "host.user.id",
    type: "keyword"
  },
]
        </pre>
        </td>
      <td>
      <pre>
[
  {
    ecs: true,
    name: "event.UPDATED",
    type: "unknown"
  },
  {
    ecs: false,
    name: "host.user.id",
    type: "keyword"
  },
]
        </pre>
      </td>
    </tr>
    <tr>
      <td rowspan=2 style="border-right:3px solid black">Keep user changes if element conflicts, and remove any <b>uncustomized</b> elements if removed in update <br><br><ul><li><b>1st elem: (A B C - keep current) </li><li>2nd elem: (A A _ - remove element)</b></li></ul></b></td>

>       <td>

Reviewer 2: I think this is an overcomplication. If there's any conflict between current and target changes, we should mark the result as a conflict and simply set `merged = current` without trying to do a clever half-merge. Because:

- It would be likely incorrect anyway.
- It would require us to implement and maintain some clever code.

I'd like to go through the examples in this table together over zoom to give a quick idea how I think they all should be processed by a diff algorithm.

Reviewer 4: I agree we could first implement a simpler predictable merged field logic, and then improve it in the following iteration.

      <td>
      <pre>
[
  cluster_one:logs-*,
]
        </pre>
      </td>
    </tr>
    <tr>
      <td>
      <pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "unknown"
  },
  {
    ecs: true,
    name: "process.executable",
    type: "string"
  }
]
        </pre>
      </td>
      <td>
        <pre>
[
  {
    ecs: true,
    name: "event.UPDATED",
    type: "unknown"
  }
]
        </pre>
      </td>
      <td style="border-right:3px solid black">
      <pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "keyword"
  },
  {
    ecs: true,
    name: "process.executable.ip",
    type: "ip"
  }
]
        </pre>
        </td>
      <td>
      <pre>
[
  {
    ecs: true,
    name: "event.UPDATED",
    type: "unknown"
  }
]
        </pre>
      </td>
    </tr>
  </tbody>
</table>

##### MITRE ATT&CK framework (`threat` field)

The `threat` field has a specific [schema](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.gen.ts#L250) that can be handled with its own algorithm in case of conflicts, in order to try to obtain the most reasonable merge between versions. 

> **TODO: write algorithm specific to this field**

Reviewer 2: I would suggest to exclude this from the RFC. A diff algo for this field will probably be the most complex and difficult to write concrete diff algorithm. I'd suggest to try this only after we have a few basic concrete algorithms already implemented.

Reviewer 4: I agree, also it is a bit strange for the users to be modifying name of technique alone. I think of this objects as bundles.

Author: Removing whole MITRE section for now

The current UI of the **Rule Updates** page has an button with the label "Update all" which currently simply updates all rules to their `target` version:

![](2024-02-23-14-09-16.png)

We can replace this button for an **Update rules with no conflicts** button. Clicking this button would make a request to `/upgrade/_perform` with a payload that includes only rules that have no conflicts. That payload can be built out of the response of the `/upgrade/_review` endpoint, including only rules with no conflicts and specifying that all rules should be updated to their `MERGED` version, which represents the "clean" (conflictless) update.

```ts
// [PSEUDOCODE]
const reviewUpgradeResponse: ReviewRuleUpgradeResponseBody = await fetch(
  '/internal/detection_engine/prebuilt_rules/upgrade/_review',
  { method: 'POST' }
);

// Filter out rules with conflicts, and create payload pointing all to 'MERGE' version
const performUpgradePayload = {
  mode: 'SPECIFIC_RULES',
  rules: reviewUpgradeResponse.rules
    .filter((rule) => !rule.diff.has_conflict)
    .map((rule) => ({
      rule_id: rule.rule_id,
      revision: rule.revision,

>       version: rule.version,

Reviewer 2: Shouldn't we be passing the target version instead of the current one, according to https://github.com/elastic/kibana/issues/177852#issuecomment-1976229816?

Author: Yes, thanks, fixed. Also `rule.version` does not exist in the response

> See [Three-Way-Diff Component ticket](https://github.com/elastic/kibana/issues/171520) that details UI for solving conflicts.

> See [designs](https://www.figma.com/file/gLHm8LpTtSkAUQHrkG3RHU/%5B8.7%5D-%5BRules%5D-Rule-Immutability%2FCustomization?type=design&mode=design&t=LkauhLzUKUatF6cL-0#712870904).

## Other changes to UX/UI

---- end of RFC -----

## Scenario for bulk accepting updates

- At rules table level
  Bulk accept all rule updates that have no conflicting fields at all → This would make the request to update all these rules with no conflicts at all.
  Accept all rule updates for non-customized rules.

- At rule (filed-per-field) level
  Accept all fields that have no conflicting changes, but still requires the user to manually solve fields with conflicts. → Once all fields are solved, the user can make the request to update the rule.
  Accept all fields that are not customized

- User Config
  Create a user config/setting that the user can turn on to accept updates automatically if there are no conflicts.
  Or we should more concretely how we want this setting to behave

> ## Other open questions

Reviewer 2: Should this text be removed or you're planning to finalize it?

Author: Removed

#### `type` subfield

The `type` subfield serves as the discriminat to the disciminated union field `rule_source` and can take two values, as explained above: `'internal'` and `'external'`.

#### `is_customized` subfield

The `is_customized` field will be a boolean field that determines whether an external/Prebuilt Rule has been customized by the user, i.e. if any of its fields have been modified and diverged from the base version of the rule, which is the version that is installed from the Prebuilt Rules `security_detection_engine` Fleet package.

This means that the `is_customized` subfield only makes sense and will be used for prebuilt rules, or rules whose `rule_source` field is of type `'external'`.

For prebuilt rules, the `rule_source.is_customized` value will be initially set to `false` when a brand new rule is installed, but will be rewritten to `true` if a rule's field is edited and diverges from the value from the base version of the rule.

See section [Updating the `isCustomized` flag](#updating-the-is_customized-field)

> #### `source_updated_at` subfield

Reviewer 3: What is `source_updated_at` field for? What problems doesn this field solve? Why it can't be retrived from `security-rule` saved object? For detections-as-code how can it be set if nno such data is provided via the API request data?

Author: Its use case is detailed in th elink porvided at the end of the paragrah. Particulary, this comment: https://github.com/elastic/detection-rules/issues/2826#issuecomment-1820876872

> Why API endpoints can't retrive it from security-rule saved object (when necessary) at runtime and return in the response?

Because the data that we need is not stored in the `security-rule` SO itself. We need the TRaDE team to add it to the object, either directly in the TOML files when they are manually edited or at buildtime.

The `source_updated_at` will be a field containing a date in ISO 8601 format which describes the last time that an Elastic Prebuilt Detection rule was created and subsequently updated by the TRaDe team, the team responsible for creating detection rules. Its usage is detailed in this [ticket](https://github.com/elastic/detection-rules/issues/2826).

This field will be optional in both the API schema and the internal rule schema, since this value will not exist for prebuilt rules until a new version of each rule which includes this field in the prebuilt rule asset is published by the TRaDE team, and the user installs it or updates to it.

### Changes needed in rule schema

As detailed further down in the [Plan for carrying out migrations of rule SOs](#plan-for-carrying-out-migrations-of-rule-sos) section, we will be performing incremental migration-on-write on our saved objects to get to our new schemas (see details in linked section).

This means that we need to differentiate between changes to the internal rule schema and the API schema.

> #### API schema

Reviewer 2: Again, it's rather hard to navigate a deeply nested hierarchy of headers. I'll post a few comments below with suggestions to simplify it an make more flat.

Let's first remove this `#### API schema` header.

> ##### API request and response rule schema

Reviewer 2: ```suggestion
#### API request and response rule schema
```

**In the API schema** the `rule_source` will be a required, as it will allow us to differentiate between Elastic prebuilt rules, and custom rules. 

Notice, as well, that `immutable` and `rule_source` will continue to be part **only of the API response schema**, and never form part of the **request parameters**.

The OpenAPI schema will need to be modified so:

_Source: [x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.schema.yaml](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.schema.yaml)_

```yaml
#  [... file continues above...]

# Add deprecation warning to `immutable` field
IsRuleImmutable:
  type: boolean

>   description: '[DEPRECATION WARNING - This field is deprecated and will be removed in a future release. Use the prebuilt field to determine if a rule is an Elastic prebuilt rule or a custom rule.] - Determines whether the rule is immutable (i.e. cannot be modified by the user).'

Reviewer 2: ```suggestion
  description: '[DEPRECATION WARNING! This field is deprecated and... <the warning explains the deprecation period and suggests a replacement> ] - Determines whether the rule is a prebuilt Elastic rule.'
```

IsExternalRuleCustomized:
  type: boolean
  description: Determines whether an external/prebuilt rule has been customized by the user (i.e. any of its fields have been modified and diverged from the base value).

ExternalSourceUpdatedAt:
  type: string
  format: date-time

>   description: The date and time that the external/prebuilt rule was last updated in its source repository.

Reviewer 2: Could we please make a note in the RFC that we're not going to add this field to the schema in the initial implementation because of the dependency on TRADE?

  oneOf:
    - $ref: '#/components/schemas/ExternalRuleSource'
    - $ref: '#/components/schemas/InternalRuleSource'
#  [... file continues below ...]
```

_Source: [x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml)_

```yaml
#  [... file continues above...]

ResponseFields:
  type: object
  properties:
    id:
      $ref: './common_attributes.schema.yaml#/components/schemas/RuleObjectId'
    rule_id:
      $ref: './common_attributes.schema.yaml#/components/schemas/RuleSignatureId'
    immutable:
      $ref: './common_attributes.schema.yaml#/components/schemas/IsRuleImmutable'
    rule_source:
      $ref: './common_attributes.schema.yaml#/components/schemas/RuleSource'
    #  [...  more response fields ...]
  required:
    - id
    - rule_id
    - immutable
    - updated_at
    - updated_by
    - created_at
    - created_by
    - revision
    - rule_source
#  [... file continues below...]
```

> ##### Rule Import request schema

Reviewer 2: ```suggestion
#### Rule Import request schema
```

See the detailed explanation for this mechanism in the [Exporting and importing rules](#exporting-and-importing-rules) sections.

The `immutable` and `rule_source` fields will be ignored if passed in the request payload.

_Source: [x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts)_

```ts
// [... file continues above...]
import {
  // [...]
  RuleSignatureId,
  IsRuleImmutable,
  ExternalSourceAttributes,
  RuleVersion,
} from '../../model/rule_schema';

export type RuleToImport = z.infer<typeof RuleToImport>;
export type RuleToImportInput = z.input<typeof RuleToImport>;
export const RuleToImport = BaseCreateProps.and(TypeSpecificCreateProps).and(
  ResponseFields.partial().extend({
    rule_id: RuleSignatureId,
    version: RuleVersion,
  })

> );

Reviewer 2: `ResponseFields` is going to include both `immutable: IsRuleImmutable` and `rule_source: RuleSource.optional()`, according to changes in this PR. These two fields should NOT be included in the `RuleToImport` schema. This type definition will need to be adjusted somehow.

Reviewer 2: @jpdjere You are right, from that perspective it should be fine to keep these fields in the `RuleToImport` schema and just ignore the input and recalculate them inside the import endpoint.

```

#### Internal rule schema

**The internal rule schema** needs to represent that the new `ruleSource` field may not always exist, so both need to be optional.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts)_

```ts
export type BaseRuleParams = z.infer<typeof BaseRuleParams>;
export const BaseRuleParams = z.object({
  // [...]

>   immutable: IsRuleImmutable,

Reviewer 2: Should this become optional in the internal schema? Will newly created rules have this param stored in ES?

Reviewer 2: @jpdjere I'm not strongly against this, but I'm not sure I understand why we'd need to store `immutable` in those rules that have been migrated-on-write or newly created. We will have normalization-on-read + fallback to `immutable` in the `find` endpoint + the field will be deprecated once we release prebuilt rule customization. If you suggest to store it anyway, then when do you think we will stop storing it?

Author: My original idea was: "remove it only when we release a new API version".

But now I think that what you propose is actually tidier and takes care of the problem now instead of later (like in two years, or so.)

Author: - Reflect this decision on the Migration-on-write section
- Update the table that shows migration scenarios

In the internal rule schema, there are two additional important reasons why we need to make sure that these two fields optional:

- When rules are executed, a call to the method `validateRuleTypeParams` is done, which is a method that validates the passed rule's parameters using the validators defined in `x-pack/plugins/security_solution/server/lib/detection_engine/rule_types`, within each of the rule query types files (for [EQL rules](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/eql/create_eql_alert_type.ts#L27), for example). The validation is done based on the internal rule schema `BaseRulesParams` displayed above. Having `ruleSource` as required field would cause custom rules or prebuilt rules that haven't had their schema migrated to fail during runtime.
- The Rule Client `update` method also calls the `validateRuleTypeParams` to validate the rule's params. Since the Rule Client's `update` method is used in our endpoint handlers, such as and `/rules/patch` and `/_bulk_actions`, these would fail when executed against a payload of custom rule.

> ##### Prebuilt Rule asset schema

Reviewer 2: ```suggestion
#### Prebuilt Rule asset schema
```

> The `PrebuiltRuleAsset` type needs to be updated to include the new `source_updated_at` date that will be progressively shipped with new versions of rules in the Elastic Prebuilt Rules package.

Author: Added 👍

> The `PrebuiltRuleAsset` type needs to be updated to include the new `source_updated_at` date that will be progressively shipped with new versions of rules in the Elastic Prebuilt Rules package.

Reviewer 2: @jpdjere I don't see any updates in this section, can you point to the place where you mentioned this?

Notice that this field will be a **top-level field in the Prebuilt Rule asset** schema, but will be part of the `rule_source` field in the API rule schema and internal rule schema.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts)_

```ts
export const PrebuiltRuleAsset = BaseCreateProps.and(TypeSpecificCreateProps).and(
  z.object({
    rule_id: RuleSignatureId,
    version: RuleVersion,
    related_integrations: RelatedIntegrationArray.optional(),
    required_fields: RequiredFieldArray.optional(),
    setup: SetupGuide.optional(),
    source_updated_at: SourceUpdatedAt.optional(), // new optional field
  })
);
```

> ### Deprecating the `immutable` field

Author: Done, thanks for the suggestion 👍

This "overuse" of these 3 methods for a variety of user actions makes their logic tightly coupled and creates a considerable amount of complexity to CRUD functions that should remain logically simple.

For example: the `createRules` method is used in 3 use cases:
1. when creating custom rules with our Rule Creation endpoints
2. when importing rules, when the imported `rule_id` does not already exist in Kibana 
3. when upgrading rules, if a rule undergoes a `type` change, the existing rule is deleted a new one is created.

The same happens with `patchRules`. It is used:
1. when patching rules via our Rule Patch endpoints
2. when importing rules, when the imported `rule_id` is brand new
3. when upgrading rules, if a rule maintains its `type`.

This causes these 3 CRUD functions to have an unnecessary complex interfaces and logic in order to handle all these different use cases.

As part of the epic, we should move away from this practice and create new CRUD methods that handle the different cases mentioned above specifically. This will help simplify our existing methods, keep logic in all of our CRUD methods simple and logically uncoupled from one another.

> ### Migration strategy

Reviewer 3: Have you considered any alternatives to the proposed approach? Obviously having `normalization on read` and `migration on write` will make codebase more complex and error prone. We'll need to test all the edge cases which require enough team resources. Has it been confirmed `Model Version API` won't support encryption in any foreseeable future?

For example, as an alternative we could consider the following

- build an additional abstracted data layer on top of `rulesClient.create()`, `rulesClient.update()` etc to make sure we always interact with mighrated rules. This layer will perform `migration on write` and `normalization on read` but it will be incapsulated. It means rule's data returned by this layer will only contain `rule_source` while `immutable` field won't be there.

- run a task manager's task to migrate rules.

Author: I indeed considered an alternative approach similar to running a backround process: migrating all rules when calling the "find all rules" endpoint. But did a couple of PoC and the performance implications were too high; API requests would timeout when large number of rules (~10k) were attempted to be migrated. 

And unluckily, there'e currently no mechanism in place using the task manager or similar to migrate SO in the background, while at the same time blocking the user from interacting with the rules while this takes place (which we need to do to maintain data consistency)

> build an additional abstracted data layer on top of rulesClient.create(), rulesClient.update()

This proposal sounds similar to what's described in the RFC. Not sure what you have in mind, but the idea is to always interact with migrated rules through the user of normalization on read and migration on write. Maybe the difference is that that logic is on the Security Solution side in this proposal, and maybe you are thinking of including it on the Alerting Framework's side? Let me know.

Author: - Create ticket for refactoring our CRUD utility functions, encapsulating them into something like a Rule Management client where the migration-on-write and normalization-on-read logic can be encapsulated.

_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts)_

```ts
export const internalRuleToAPIResponse = (rule) => {
  return {
    ...commonParamsCamelToSnake(rule.params), // <--- rule params are converted here
    ...typeSpecificCamelToSnake(rule.params),
    // [... more object properties ...]
  };
};

export const commonParamsCamelToSnake = (params: BaseRuleParams) => {
  const { immutable, rule_source } = normalizeRuleSourceSchemaOnRuleRead(params);

  return {
    immutable,
    rule_source,
    // [... more object properties ...]
  };
};
```

And the `normalizeRuleSourceSchemaOnRuleRead` can be defined so:

_Source: x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/prebuilt_rule_schema_migrations.ts_ (New file)

```ts
interface OnReadNormalizationResult {
  immutable: IsRuleImmutable;
  rule_source: RuleSource
}

/**

>  * To calculate `rule_source`:
>  * - Use `rule_source` field  if already exists in the internal rule object params.
>  * - If it does not exist, check the value `immutable` from the rules param:
>  *   - if it is `true`, create a new `rule_source` object, with `isCustomized` set to `false` and `repoNam` set to `elastic_prebuilt` (Use case of rule_source rules that have not been yet migrated-on-write)
>  *   - otherwise, the field should be `undefined` (use case of custom rules).

Reviewer 2: This comment is not correct anymore, although I'm not sure if we should have it because the code is self-explanatory. Anyway:

```suggestion
 * To calculate `rule_source`:
 * - Use `rule_source` field  if already exists in the internal rule object params. This means the rule has been migreted-on-write.
 * - If `rule_source` does not exist (the rule hasn't been migrated yet), check the value of `immutable`:
 *   - if it is `true`, create a new `rule_source` object of type `external`, with `is_customized` set to `false`
 *   - if it is `false`, create a new `rule_source` object of type `internal`
```

Author: Relying on code, removing text explanation

 */
const getRuleSourceValueForRuleRead = (ruleParams: BaseRuleParams): RuleSource | undefined => {
  if (ruleParams.rule_source) {
    return ruleParams.rule_source;
  }

  if (ruleParams.immutable) {
    return {
      type: 'external'
      isCustomized: false,
    };
  }

  return {
    type: 'internal'
  };
};

export const normalizePrebuiltSchemaOnRuleRead = (
  ruleParams: BaseRuleParams
): OnReadNormalizationResult => {

/**
 * To calculate `immutable`:
 * - Checks if the `rule_source` field exists in the rule's parameters.
 *   - If `rule_source` exists and its type is external, sets `immutable` to `true`. (Use case of Rules that have already been migrated-on-write)
 *   - If `rule_source` does not exist, return the value of the params' `immutable` field. (Use case of Rules that have not yet been migrated on write.)
 */
  const isRuleSourceExternal = ruleParams.rule_source && ruleParams.rule_source.type === 'external';
  const immutable = Boolean(isRuleSourceExternal) || Boolean(ruleParams.immutable);
  const rule_source = getRuleSourceValueForRuleRead(ruleParams);

  return {
    immutable,
    rule_source,
  };
};
```

> ##### Rule Management endpoints that will perform normalization-on-read

Reviewer 3: nit: Having a table like

| API endpoint | migration type | comments |

could improve readability.

Reviewer 2: ++ I think it's a great idea

Author: Added a table and removed much of the text. I think it's much clearer now.

> - All endpoints that also perform migration-on-write, as listed in the below `Migration on write` section. All of them use the `internalRuleToAPIResponse` methods to transform the rule before returning the response, so the normalization on read will take place there. For these specific cases, however, the normalization-on-read will take place, but the data stored in ES for each rule should match the response given by the endpoint, i.e. the normalized in-memory value and the stored value of the rule should be the same.

Reviewer 3: The text here looks overcomplicated. I'd recommend to split this section in two sub parts read API endpoints and delete/bulk endpoints. The latter return rule's data and need to make sure returned data is normalized.

Btw, what's the problem to perform migration-on-write when enabling/disabling rules? Could you elabored on the reasons in the RFC?

Author: Bulk enabling and disabling rules use the RuleClient's `enable` and `disable`  methods. We pass in only the `id` and it fetches the attributes and does the modification of the `enabled` prop.

Doing migration-on-write is techninally possible, but would mean adding the migration-on-write logic on the RuleClient side, which I would prefer to avoid. I'd like to keep all migration logic on the Security solution side. 

On the other side, enabling and disabling are probably the most-often executed actions, so if we want to accelerate the migration of a user's SO, maybe we should actually do this.

What do you think? cc @banderror

Reviewer 2: > Doing migration-on-write is techninally possible, but would mean adding the migration-on-write logic on the RuleClient side, which I would prefer to avoid.

@jpdjere I also think we should avoid this. The RulesClient doesn't and should not know about our rule params. Properly implementing rule params migrations on enable/disable would require creating a special mechanism for that, which doesn't seem to worth it.

Besides that, we're going to migrate to the bulk enable/disable methods of the RulesClient very soon: https://github.com/elastic/kibana/issues/177634.

Author: - Detail in RFC why not doing migration-on-write on bulk enabling/disabling.
- Create a separate ticket for evaluating migration-on-write on enabling/disabling for small amounts.

- **Find Rules - `GET /rules/_find`**
- **Read Rule - GET /rules:**
- **Delete Rules - DELETE /rules**

> - **Bulk Actions** - `POST /rules/_bulk_action`: all other action types not mentioned in the previous section will perform normalization-on-read before responding:

Reviewer 3: What previous section do you refer here? Did I get it right only `Enable`, `Disable`, `Delete` and `Export` perform migration-on-write?

Reviewer 2: What's special about the Enable, Disable, Delete and Export bulk actions? Shouldn't just all actions perform normalization-on-read?

Author: Yes, sorry, this is bad redaction. All bulk action types will perform normalization on read. The list here tried to explain that Enable, Disable, Delete and Export ONLY do normalization-on-read, and NOT migration-on-write.

I'll replace all of this with a table anyways.

Reviewer 2: @jpdjere The table looks much better 👍

  - Enable
  - Disable
  - Delete
  - Export: see Importing and Exporting Rules section below
- **Export Rules** - `POST /rules/_export`:
  - this endpoint is unused via the UI, but might still be used via the public API
  - same logic as for `_bulk_action` with the `export` action explained in section [Exporting rules](#exporting-rules) below: only normalization-on-read will be performed before writing to the output `ndjson` file.

#### Migration on write

This type of migration will be in charge of updating saved object fields in Elasticsearch from the legacy form to the new form, and will be performed on all write operations of the rules (see list below). This means that we need to perform this type of migration of rules whenever any endpoint operation that writes/updates a rule saved object is called.

> > In Elasticsearch, the type of rule saved objects is called `alert`.
> 
> This migration strategy of write/update operations means that a user's data will be migrated incrementally, and that both types of data (non-migrated and migrated saved objects) will coexist for an indeterminate amount of time. Therefore we have to maintain backwards data compatibility for the non-migrated data types.

Reviewer 2: ```suggestion
This migration strategy means that rules will be migrated incrementally, and that both non-migrated and migrated rules will coexist for an indeterminate amount of time. Therefore we have to maintain backwards compatibility with non-migrated rules.
```

Author: Applied

The migration logic should take place within the handler logic of all endpoints that carry out write/update endpoint operations, and the endpoint should return the already-migrated rule(s).

##### Rule Management endpoints that should include migration-on-write logic

> All of the following endpoints either fetch the rule before updating it, or send the rule's params as part of the body in the request. Therefore, we can apply migraton logic to them -as described below- before saving the rules to Elasticsearch.

Reviewer 3: typo `migraton` -> `migration`

  - See section [Importing rules](#importing-rules)
- **Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal): current way of upgrading a prebuilt rule
- **(LEGACY) Install Prebuilt Rules And Timelines** - `PUT /rules/prepackaged`: legacy endpoint for installing prebuilt rules. Also updates existing prebuilt rules package, so these rules should be migrated during update as well.
- **Bulk Actions** - `POST /rules/_bulk_action`:
  This endpoint includes many types of actions, but only the **bulk edit** and the **duplicate** actions should perform a migration on write.

This endpoint also includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`, since we never write to ES when this parameter is set to `true`.

For the **bulk edit** action, we can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed. See implementation details in the below [Bulk editing rules](#bulk-editing-rules) section.

For the **duplicate** rule action:

> Since we will be creating a new rule on ES, we should create it with the new schema. Per definition, all duplicated rules will be `custom` rules. That means that for all rules -including prebuilt rules-, when duplicating, the `rule_source` field will not be created on the newly created rule.
> This action will not perform a migration-on-write of the original rule being duplicated for two reasons:

Reviewer 2: ```suggestion
Since we will be creating a new rule on ES, we should create it with the new schema. Per definition, all duplicated rules will be `custom` rules. That means that all duplicated rules (the duplicates) should get a `rule_source` of type `internal`.
This action will not perform a migration-on-write of the original rule being duplicated for two reasons:
```

> - it would violate the principle of least surprise for the endpoint

Reviewer 3: It'd be nice to have the reasons highlighted before deep diving into each endpoint. It will also explain why normalization-on-read is used while ut could be migration-on-write.

Let's see all possible use cases that will require migration-on-write, the endpoints that they apply to, and the expected resulting migrated field, based on the action and their input.

#### Updating and upgrading rules

Updating rules can currently be performed via five endpoints:

- **Update Rule** - `PUT /rules`
- **Patch Rule** - `PATCH /rules`
- **Bulk Update Rules** - `PUT /rules/_bulk_update`
- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`
- **Bulk Actions** - `POST /rules/_bulk_action`: with **bulk edit** action

Upgrading prebuilt rules to their newer version is done by two endpoints:

- **(LEGACY) Install Prebuilt Rules And Timelines** - `PUT /rules/prepackaged`
- **Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal)

The legacy endpoint does not allow for customization of fields during the upgrade, but the new rule upgrade customization endpoint does. 

Additionally:

- **Bulk Actions** - `POST /rules/_bulk_action`: with **duplicate** action

> will perform migration but does not allow for customization during the duplication proces.

Reviewer 3: type `proces` -> `process`

Why should Bulk duplicate allow customization? Or the meaning of this sentence is to highlight that Bulk duplicate doesn't accept any params besides rule ids?

So we can analyze the expected outputs of the migration of all these 8 endpoints together.

The resulting values for `immutable` and `rule_source` when calling these endpoints, and the migration being performed in the background, should be as follows:

<table>
  <thead>
    <tr>
      <th>Migration use case</th>
      <th>Current value of <pre>rule_source</pre></th>
      <th>Current value of <pre>immutable</pre></th>
      <th>Any field diverges <br> from base version after update?</th>
      <th>Results</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Custom rule</b> (not migrated yet)</td>
      <td>
        <pre>undefined</pre>
      </td>
      <td><pre>false</pre></td>
      <td>N/A - Doesn't apply for custom rules</td>
      <td>
        <pre>
          {
            rule_source: {
              type: 'internal'
            },
            immutable: false,
          }
        </pre>

>       </td>

Author: Thanks, fixed!

      <td>Yes</td>
      <td>
        <pre>
          {
            rule_source: {
              type: 'external',
              isCustomized: true,
              ...
            },
            immutable: true,
          }
        </pre>
      </td>
    </tr>

  </tbody>
</table>

As explained before, all these endpoints suffer from the tightly-coupled logic problem described in the section [`Problem with tightly coupled logic in our endpoints`](#problem-with-tightly-coupled-logic-in-our-endpoints).

We should therefore create new CRUD methods as needed to uncouple them logically and cleanly apply migrations to them.

---

#### Bulk editing rules

The endpoint **Bulk Actions** - `POST /rules/_bulk_action` has the same outputs from the migration logic as the endpoints listed in the above section, so it was included in the scenarios in the table above. 

However, this endpoint has some specific details that should be mentioned.

Firstly, when bulk editing rules, the migration should be carried out for the following use cases:
  - Bulk adding or deleting index patterns
  - Bulk adding or deleting tags
  - Updating rule schedules
  - Adding rules actions

>   - Duplicating a rule (with the migration applied only to the new rule)

Author: Yes, I mean here just creating the `rule_source` property in the newly created rule only, with the value that you mention.

Out of the actions mentioned above, the only use cases that should possibly result in the migration having a result of `rule_source.isCustomized = true` are the first three:
  - Bulk adding or deleting index patterns
  - Bulk adding or deleting tags
  - Updating rule schedules

> That means that updating a rule's actions or duplicating a rule should not be considered a customization of a prebuilt rule.

Author: Nothing to elaborate actually, just stating the obvious: since we are generating a custom rule, "customization" is not possible/doesn't make sense. 

I'll remove the reference to "duplicating a rule" to avoid confusion,

Secondly, in contrast with other endpoints, the migration for bulk editing rules needs to be carried within the Alerting Framework's `RuleClient` since we need to migrate the rule's `rule_source` and `immutable` parameters before the `RuleClient` does the saving of the rule into Elasticsearch.

This however, gives us the advantage that the `RuleClient` already includes logic to calculate if a rule's attributes or parameters have been modified from its initial values, so we can rely on that to calculate the `rule_source.is_customized` field during the update.

> The `RulesClient` class has a `bulkEdit` method, which is called by our **Bulk Actions endpoint** `POST /rules/_bulk_action`. That method includes complex logic, but we can focus on the `updateRuleAttributesAndParamsInMemory` method, where the rule attributes and parameters are updated before being saved to ES. This function also calculates the booleans `isAttributesUpdateSkipped` and `isParamsUpdateSkipped` which we can leverage to calculate the new value of the `rule_source.is_customized` field in our params.

Author: Yes, that's true. But here I'm referring to the calculation of the `is_customized` field. The update of fields does happen within the Alerting Framework, and if we want to know if a value was indeed updated so that it has diverged from its base version, we need to that check within the logic of the Rules Client.

> The `RulesClient` class has a `bulkEdit` method, which is called by our **Bulk Actions endpoint** `POST /rules/_bulk_action`. That method includes complex logic, but we can focus on the `updateRuleAttributesAndParamsInMemory` method, where the rule attributes and parameters are updated before being saved to ES. This function also calculates the booleans `isAttributesUpdateSkipped` and `isParamsUpdateSkipped` which we can leverage to calculate the new value of the `rule_source.is_customized` field in our params.

Reviewer 2: After looking into the code, I think we should do it here, somewhere inside the `paramsModifier` callback:

https://github.com/elastic/kibana/blob/a7c8ef421ee097810af51413b621871874becfa4/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/bulk_edit_rules.ts#L30-L62

But not inside the `updateRuleAttributesAndParamsInMemory` on the framework side.

> The `RulesClient` class has a `bulkEdit` method, which is called by our **Bulk Actions endpoint** `POST /rules/_bulk_action`. That method includes complex logic, but we can focus on the `updateRuleAttributesAndParamsInMemory` method, where the rule attributes and parameters are updated before being saved to ES. This function also calculates the booleans `isAttributesUpdateSkipped` and `isParamsUpdateSkipped` which we can leverage to calculate the new value of the `rule_source.is_customized` field in our params.

Author: The problem is that for calculating `is_customized` we need to know what the current values of the rule SO are. In the Security Solution side (i.e.: `x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/bulk_edit_rules.ts`), we just create the set of operations that modify the attributes and params of the rules and pass them over to the Alerting Framework's `bulkEdit` method.

Only in the Alerting Framework is the rule SO fetched and its params and atributes modified. 
So basically, the Security Solution side nevers sees the current values of the rule's fields. It only recieves the already modified values, in `result`.
This means that if we want to do the comparison of the current and resulting values of the rule's fields, we have two options:
1. Do the comparison in the Alerting Framework side, where the rule is fetched and the modifications applied.
2. We need to fetch the rule SO on the Security Solution side, -which currently does not happen-, and then compare the current values with the resulting values. 

Now I think that this second options is preferable, to avoid adding logic to the Alerting Framework side, but we might need to consider the performance overhead of fetching the rule SO.

> The `RulesClient` class has a `bulkEdit` method, which is called by our **Bulk Actions endpoint** `POST /rules/_bulk_action`. That method includes complex logic, but we can focus on the `updateRuleAttributesAndParamsInMemory` method, where the rule attributes and parameters are updated before being saved to ES. This function also calculates the booleans `isAttributesUpdateSkipped` and `isParamsUpdateSkipped` which we can leverage to calculate the new value of the `rule_source.is_customized` field in our params.

Author: Calculation of `is_customized` options:
1. Modify the rulesParamsModifier to take as parameter the whole rule, or additionally the operations.
2. Do it as part of the `validateMutatedRuleTypeParams` method `x-pack/plugins/alerting/server/lib/validate_mutated_rule_type_params.ts`

Here, we need to update the KQL filters and the logic for fetching Elastic prebuilt and custom rules, relying on `rule_source` but with fallback to `immutable`:

_Source for `convertRulesFilterToKQL`: [x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts)_

In order to retrieve Elastic prebuilt rules, we can filter for rules in which `alert.attributes.params.rule_source.type` is `external`, with a fallback to rules in which `alert.attributes.params.immutable` is `true`, for rules which have not had their schema migrated and `rule_source` does not still exists for them.

> In the future, we can further distinguish between `type: external` rules by their source origin. If we introduce a new field within `rule_source` that defines their origin (such as `repo_name`, `repo_id`, or similar), we can further filter by Elastic prebuilt rules, or rules that are externally sourced from a repository handled by the user, etc.

> 

Reviewer 4: I think we need to accommodate filtering for "Elastic prebuilt customised rules" too.

This endpoint will require major changes to add the capability of letting users selecting a custom version of the rule with custom values for rule fields. This will be explained further below in the "Changes to upgrade `_review` and `_perform` endpoints" section.

The calculation of the value for the `rule_source.is_customized` field for the updated rule will depend on that logic as well, so its calculation will be explained in that section.

Again, this endpoint suffers from tightly coupled logic explained in [`Problem with tightly coupled logic in our endpoints`](#problem-with-tightly-coupled-logic-in-our-endpoints): it uses the `upgradePrebuiltRules` method to actually upgrade the rules, but this method either patches existing rules -for the normal case-, or deletes an existing rule and recreates it if the rule underwent a type change, using the `patchRules` and `createRules` methods respectively. We should decouple that logic by introducing a new CRUD method with a specific use case for upgrading prebuilt rule.

> The changes in the `upgradePrebuiltRules` method need to take into account both paths. In the endpoint logic handler, when calculating if the rule was customized by the user, we will create a boolean called `isRuleCustomizedDuringUpgrade`, that can be passed as an argument to `upgradePrebuiltRules`:

Reviewer 2: This is probably too much assumptions about insignificant implementation details for an RFC.

Author: Removed

Also, in order to allow the endpoint to export both custom **and** prebuilt rules, we need to update the logic and remove the checks that we currently do server-side in both of these methods, and which filter out prebuilt rules from the response payload:

_Source for `getRulesFromObjects`: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_by_object_ids.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_by_object_ids.ts)_

_Source for `getExportAll`: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts)_

### Importing rules

The user will now be able to import both custom and prebuilt rules (including any customizations that might have been performed on its modifiable fields).

We want to handle the case of importing both migrated and non-migrated rules via an `ndjson` file.

> If a user imports a prebuilt rule, Kibana should continue to track the rule asset from the `security_detection_engine` package if the `rule_id` matches. This means that a user will be able to update a previously exported prebuilt rule, visualize any diffs on any customized fields, and continue to receive updates on the rule.

Author: Applied

To allow for importing of Elastic prebuilt rules, we will **not rely** in the `rule_source` or the legacy `immutable` fields (which are not part of the import endpoint parameters), but we will rather **calculate them dynamically based on the `rule_id` and `version` request parameters**.

The logic to importing a rule is as follows:

>  - First, read the import endpoint request parameters `rule_id` and `version`. These two are the two required parameters in the endpoint, while all other rule fields are optional.

Reviewer 3: What about a minimum required set of rule fields? For example a custom query can't have only `rule_id` and `version`.

Author: Yes, the RuleImport schema specifies which rule fields are required:
```ts
export const RuleToImport = BaseCreateProps.and(TypeSpecificCreateProps).and(
  ResponseFields.partial().extend({
    rule_id: RuleSignatureId,
    version: RuleVersion,
  })
);
```
This is defined in the "Rule Import request schema" section above

- Secondly, check that the `security_detection_engine` Fleet package is installed and install it if it isn't. We will need the package to be installed to check if a rule that is being imported is an Elastic prebuilt rule.

- Then, using the `rule_id` and `version`, attempt to fetch the corresponding `security-rule` asset from ES. 

- **If a matching `rule_id` and `version` is found**, that means that the rule is an Elastic prebuilt rule, and we should therefore dynamically calculate the rule's `rule_source` field and its subfields:
  - `type`: should be always `external` since a matching external `security-rule` was found.
  - `source_updated_at`: can be retrieved from the corresponding `security-rule` asset.

>   - `is_customized`: should be calculated based on the `security-rule` asset's field and the field rules that were part of the import request parameters. If any of them are different, i.e. have diverged from the base version, `is_customized` should be true.

Author: Applied

- Finally, using the import payload, plus the rule's `security-rule` asset fields and the calculated `rule_source` fields, create the rule, or update it if already exists in Kibana.

> **If a matching `rule_id` is found, but the `version` is not found**, reject the import for that rule. Without a version, we cannot dynamically calculate `rule_source`.

Reviewer 2: Another argument for NOT rejecting the import in this case is the fact that the prebuilt rules package has a limit on the number of historical rule versions, and we can't assume that for a given rule_id we will always have ALL historical versions as `security-rule` assets.

> **If a matching `rule_id` is found, but the `version` is not found**, reject the import for that rule. Without a version, we cannot dynamically calculate `rule_source`.

Author: Thanks, this makes a lot of sense. I corrected this scenario, with the explanation.

> **If a matching `rule_id` is found, but the `version` is not found**, reject the import for that rule. Without a version, we cannot dynamically calculate `rule_source`.

Reviewer 2: @jpdjere Thank you. One last thought: in order to be consistent with this edge case of calculating `is_customized`:

<img width="1624" alt="Screenshot 2024-03-27 at 14 06 55" src="https://github.com/elastic/kibana/assets/7359339/153e6b79-574a-4513-97d7-000b3a58d716">

should we set `{ rule_source: { type: 'external', is_customized: true }` on import when a matching `rule_id` is found, but the `version` is not found?

  source_updated_at: "2024-05-..."
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My prebuilt rule",
} 
</pre>
    </td>
    <td>N/A</td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: false,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
      </td>
    <tr>

  <tr>
    <td>
      Not installed
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  source_updated_at: "2024-05-..."
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
} 
</pre>
    </td>
    <td>N/A</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: true,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
      </td>
    <tr>

  <tr>
    <td>
<pre>
{
  name: "My prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: false,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  source_updated_at: "2024-05-..."
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
} 
</pre>
    </td>
    <td><b>No</b></td>
    <td>

>     Rule import rejected<br> because of rule_id match

Reviewer 3: Is rule import rejected because of rule overwrite flag is set to false?

Author: Exactly. Added that to the message in the last column

      </td>
    <tr>

  <tr>
    <td>
<pre>
{
  name: "My prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: false,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  source_updated_at: "2024-05-..."
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: true,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
      </td>
    <tr>

  <tr>
    <td>
<pre>
{
  name: "My custom prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: true,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
    </td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  source_updated_at: "2024-05-..."
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My prebuilt rule",
} 
</pre>
    </td>
    <td>Yes</td>
    <td>
<pre>
{
  name: "My prebuilt rule",
  rule_source: {
    type: "external",
    is_customized: false,
    source_updated_at: "2024-05-..."
  }
} 
</pre>
      </td>
    <tr>

> 

Reviewer 3: It'd be nice to have a column named `Matches security-rule asses` to simpler reason about calculated `is_customized` value.

The endpoint **Bulk Actions** - `POST /rules/_bulk_action` does provide validation in the endpoint logic itself: if a user attempts to edit prebuilt rule (`immutable: true`) the endpoint rejects that edit in two ways:

- in `dryRun` mode, with an error: "editing prebuilt rules is not supported".
- in normal mode, with validation that throws the error "Elastic rule can't be edited".

In both cases, the validation checks if the `immutable` param of the rule is `false`, or if the action sent in the payload is setting or adding actions to a rule. If any of those two conditions are true, the validation succeeds and the rule(s) can be edited.

#### Changes needed to endpoints

Depending on the endpoint, we will have to modify it to address multiple changes. These are:

> **1. Introduction of migration logic**: as explained in the first section of this document, endpoints should have the additional reponsibility of migrating our rule schema.

Reviewer 3: typo `reponsibility` -> `responsibility`

Author: Fixed

> **2. Calculation of the `is_customized` field**: this field which is part of the `rule_source` field for external rules needs to be recalculated whenever a rule's field is possibly modified. See implementation details in the [Updating the `is_customized` field](#updating-the-is_customized-field) section.

Author: Fixed

<pre>
{
  name: "Original name",
  rule_source: {
    type: "external",
    is_customized: false,
    ...
  }
}
</pre>
    </td>
  </tr>

  
  <tr>
    <td>Field existing in base asset is removed</td>
    <td>
<pre>
{
  name: "Original name",
  "tags": ["Linux", "Windows"]
  rule_source: {
    type: "external",
    is_customized: false,
    ...
  }
}
</pre>
    </td>
    <td>
      <pre>
{
  name: "Original name",
  "tags": ["Linux", "Windows"]
}
      </pre>
    </td>
    <td>
      <pre>
{
  name: "Original name",
}
      </pre>
    </td>
    <td>
<pre>
{
  name: "Original name",
  rule_source: {
    type: "external",
    is_customized: true,
    ...
  }
}
</pre>
    </td>
  </tr>

  
  <tr>
    <td>Edge case: base asset not found</td>
    <td>
<pre>
{
  name: "Original name",
  rule_source: {
    type: "external",
    is_customized: false,
    ...
  }
}
</pre>
    </td>
    <td>
      <center><b>Not found</b></center>
    </td>
    <td>
      <pre>
{
  name: "Original name",
}</pre>
    </td>
    <td>
<pre>
{
  name: "Original name",
  rule_source: {
    type: "external",
    is_customized: true,
    ...
  }
}
</pre>
    </td>
  </tr>

  </tbody>
  </table>

> Notice the last scenario from the table, which deals with the edge case of the corresponding version of the `security-rule` asset not being found. This would prevent from doing a comparison of the fields to calculate if the rule should be considered customized or not.

Reviewer 3: Could you provide a few examples why `security-rule` asset isn'f found? Obviously it could be a problem with prebuilt rules installation or manually removing the asset. But will each rule update endpoint wait for prebuilt rules to be installed or not?

Author: > Could you provide a few examples why security-rule asset isn'f found?

- User manually deletes package and assets
- A user imports a prebuilt rule with matching `rule_id` but no matching `version` (base version missing)

> But what if the user then manually deleted the package, and all the assets were deleted as well?

I think that the behaviour should be to still mark the rule as `is_customized: true`, given the reasons provided below. Do you have in mind any reasons why it should behave differently?

Reviewer 2: > I think that the behaviour should be to still mark the rule as is_customized: true, given the reasons provided below.

@jpdjere This makes sense to me.

Author: TODO:
- Add edge case of importing rule from newer version of Kibana/rules package to older version, and the base version not being found cause it doesn't exist yet. -> keep the `version` from payload and mark as `is_customized: true`.

> So, if the corresponding rule asset is not found, do not attempt to do any comparisons and mark the rule as `is_customized: true`; ignoring the current values of the rules, the payload and the end result.

Author: Yes, this is something that I forgot to detail but the logic should be: "`is_customized` should be set to true if any of the fields in `BaseRequiredFields`, `BaseOptionalFields` or `BaseDefaultableFields` from the rule schema diverge from the base version". **With the exception of**:

- exceptions
- actions
- `author` and `license` (which shouldn't be customizable anyways)
- `output_index` and `namespace` (deprecated)
- `meta` (to be deprecated)

> So, if the corresponding rule asset is not found, do not attempt to do any comparisons and mark the rule as `is_customized: true`; ignoring the current values of the rules, the payload and the end result.

Author: Added to RFC

Reasons:
- A prebuilt rule which is being modified has a much higher chance of ending up as `is_customized: true` than `false`. Given that we cannot know this for certain, marking it as customized is the best educated guess.
- If we ended up making a wrong assumption in that previous point, it can be corrected the next time the rule is upgraded, or if the rule is modified again and the rule asset is found this time.

> - A rule which is missing its base version makes more sense being customized than non-customized from a domain-modelling poitn of view.

Reviewer 3: typo `poitn` -> `point`

| **Actions**    | - Actions<br>- Response actions (custom query rules only)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |

> Out of these four tabs, only **Actions** is enabled and accesible when editing a prebuilt rule - since actions (and shared exceptions lists and snoozing) are the only fields that can currently be modified for prebuilt rules from the UI.

Reviewer 3: type `accesible` -> `accessible`

All of the fields in the UI, listed above, are currently editable for custom rules, except for rule type, which is read only.

Once done editing the rule, the user clicks on the "Save Changes" button, which calls the **Update Rule** - `PUT /rules` endpoint, passing the payload for the whole rule.

**Expected behaviour for customizing prebuilt rules**

- All four tabs of the Rule Edit page should be enabled, and all the fields within each tab should be editable, as they currently are for custom rules.

> - The only field in the UI that should not be customizable is **Rule Type**, that should continue to be read-only.

Author: Updating

#### Via Bulk Actions

Custom rules can currently be updated via the Rule Table's **Bulk Actions**, which uses the **Bulk Actions** - `POST /rules/_bulk_action` endpoint.

Apart from enabling/disabling a rule - the only other action that modifies a rule's saved object and is currently possible for prebuilt rules - the user can use bulk actions to:

- Add, delete and overwrite **index patterns**
- Add, delete and overwrite **tags**
- Add and overwrite **rule actions**
- Update the rule's **schedule: interval and lookback time**
- Select a **timeline template** for the rule

As explained above, the UI validates that the five actions listed above are only possible to custom rules by using the `dryRun` mode of the **bulk actions** endpoint. If any of the selected rules in the request payload is a prebuilt rule, a message is displayed to the user informing them that modifying prebuilt rules is not possible, and that rule is removed from the subsequent request that is made in normal mode to actually edit the rule.

**Expected behaviour for customizing prebuilt rules**

- The Bulk Actions UI should now allow users to edit prebuilt rules.

> - All bulk actions actions listed above should now be applicable to prebuilt rules (This means that the changes in the API for the validation done in `dryRun` mode should not remove the rules from the payload anymore).

Reviewer 3: typo `actions actions` -> `actions`

Author: Thanks

All rules listed here can be modified in the following ways:

- enabling and disabling
- snoozing/unsnoozing or scheduling/unscheduling snooze
- updating their API keys
- deleting them

All of these actions are currently applicable for both prebuilt security rules as well as custom security rules, and they use the Alerting Framework endpoints of:

- `POST /alerting/rules/_bulk_edit`
- `PATCH /alerting/rules/_bulk_delete`

This means that **no changes will be needed in this page.**

## List of things to fix (create tickets)

1. Validation of EQL queries incorrectly fails: valid EQL queries that come with Prebuilt Rules are rejected by the EQL validator when trying to edit a rule. See [link](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/schema.tsx#L186).
2. Custom Query field overflows the viewport and cannot be completely visualized/edited when the query is too long, i.e. occupies too many lines. This is often the case from Prebuilt Rules. The editable field also does not allow scrolling. See example:

>    ![](2023-12-01-14-13-07.png)

Author: Removed

We propose developing more adaptable diff algorithms tailored to specific rule fields or rule field types. These more specific algorithms aim to minimize conflicts and automate merging of changes, doing the best effort to keep the intended customizations applied by the user, as well as the updates proposed by Elastic. Hopefully, the user will be able to simply review the proposal and accept it.

#### Concrete field diff algorithms by type

Depending on the specific field or type of field we might want to apply a specific merging algorithm when conflicts arise. Let's propose different types.

##### String fields

For string fields, in case of three-way conflicts, we will make a best effort to reconcile any modifications done by the user and any updates proposed by Elastic.

Examples: `name`, `description`, `setup`, `note` (Investigation guide)

>   <table>

Reviewer 2: I echo @maximpn's comment here, it looks like overengineering to me too.

First, I would treat single-line fields like `name` (especially if they are keywords like `severity`) and multi-line text fields like `description` and `setup` differently.

Then, for single-line fields we could use the existing simple diff algorithm:

- if `base !== current !== target`, we say it's a **conflict** and use the `current` version as the merged one
- if `base === target && base !== current`, we set `merged = current` without a conflict
- if `base === current && base !== target`, we set `merged = target` without a conflict

For multi-line text or markdown fields I think we should implement a git-like merge algorithm where changes in different areas of the text would be mergeable, but changes to the same line would render a conflict.

Finally, a couple of thoughts about single-line string fields:

- Merging keywords (especially enums) should never be done, because it could generate an invalid value.
- Changes to a rule's name might indicate some semantical changes to the whole rule (e.g. related to its source data, query/filters logic, etc). So, if both Elastic and user changed the name of the same rule, this could be an indication that the two versions of these rule are not compatible with each other, and the changes need to be reviewed by the user. Generating a conflict in this case would help the user to pay attention to all the changes.

>   <table>

Author: Splitted string diff algorithms section into two tables, simplified them and proposed a library we can use for multi-line string diffs.

>   <table>

Reviewer 2: @jpdjere The updated text looks great, thank you!

  </table>

##### Number fields

Number fields should be treated as a whole unit, i.e. not breakable by digits. Therefore, there is only one possibility of conflicts, the scenario (A B C). In that case, **keep the current version** with the user customization.

Examples: `risk_score`, `max_signals`

  <table>
    <thead>
      <tr>
        <th style="border-right:3px solid black">Use case</th>
        <th>Base version</th>
        <th>Current version</th>
        <th style="border-right:3px solid black">Target version</th>
        <th>Merged version (output)</th>
      </tr>
    </thead>
    <tbody align="center">
      <tr>
        <td style="border-right:3px solid black">Keep customizations if conflict</td>
        <td><code>50</code></td>
        <td><code>65</code></td>
        <td style="border-right:3px solid black"><code>20</code></td>
        <td><code>65</code></td>
      </tr>
    </tbody>
  </table>

##### Array fields

###### Array of strings fields

For **array of strings** fields, conflict resolution will take place on an element by element basis - that means that the algorithm will not consider or try to solve changes within a single element. For example:

```

> base: [test1]

Reviewer 2: ++, please add a few more examples with: a) different number of elements in an array and b) different kind of changes in the current and target versions

Example **array of strings** fields: `index`, `tags`, `references`,

###### Array of objects fields

Since the properties in each object within the array are highly dependent on one another, the whole object should be treated as a block, and not attempt to diff or merge changes to individual fields within each element.

For example: in a `required_fields` update, if the `type` is updated from `unknown` in the base version to `keyword` in the `target` version, but in the same object the user has updated the base `name` of the field, the change of type will have a high probability of not making sense anymore - there is no way of knowing if the change proposed by Elastic will also apply to the field name that the user is now referring through its customization.

Therefore, we should prefer to keep the user's version for the whole block.(See this example in the first element of the array in the first row of the table below).

Examples: `related_integrations,` `required_fields`

###### Proposed algorithm

- Do element by element comparison:
  - If element index exists in the three versions and the values:

>     - do not differ between all versions (base, current and target), **keep current** version. (A A A)

Reviewer 3: It could be an explanation or a link for another doc somewhere what designation `(A A A)` mean.

    - do not differ between base and current, but updates target, **keep target** version. (A A B)
    - differ between base and current, but doesn't update target, **keep current** version. (A B A)
    - differ between base and current, updates target, **keep current** version. (A B B)
    - differ between all version (base, current and target), **keep current** version. (A B C)
  - If element index exists in base and current version, but not in target:
    - If base differs from current: **keep current** version. (A B _)
    - If base equals current: **remove element**.  (A A _)
  - If element index exists in base and target version, but not in current:
    - **Remove element**. (A _ B) and (A _ A)
  - If element index exists in current and target version, but not in base:

>     - If base equals current: **add element/keep target**. (_ A A)

Reviewer 3: How can base equal to current if it doesn't exist?

Author: Revamping all logic

      <ul><li><b>1st elem: (A B C - keep current) </li><li>2nd elem: (_ A B - keep both current and target)</b></li></ul>
      </td>
      <td>
      <pre>
[
  logs-*,
]
        </pre>
      </td>
      <td>
        <pre>
[
  cluster_one:logs-*,
  new-target
]
        </pre>
      </td>
      <td style="border-right:3px solid black">
      <pre>
[
  logs-new-*,
  elastic-index
]
        </pre>
        </td>
      <td>
      <pre>
[
  cluster_one:logs-*,
  new-target,
  elastic-index
]
        </pre>
      </td>
    </tr>
    <tr>
      <td>
      <pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "unknown"
  }
]
        </pre>
      </td>
      <td>
        <pre>
[
  {
    ecs: true,
    name: "event.UPDATED",
    type: "unknown"
  },
  {
    ecs: false,
    name: "my.new.field",
    type: "keyword"
  },
]
        </pre>
      </td>
      <td style="border-right:3px solid black">
      <pre>
[
  {
    ecs: true,
    name: "event.action",
    type: "keyword"
  },
  {
    ecs: true,
    name: "new.elastic.ip",
    type: "ip"
  }
]
        </pre>
        </td>
      <td>
      <pre>

> [

Reviewer 3: This example looks "broken". Names `host.user.id` and `process.executable` aren't presented in the columns to the left.

**TODO: write algorithm specific to this field**

  <table>
    <thead>
      <tr>
        <th style="border-right:3px solid black">Use case</th>
        <th>Base version</th>
        <th>Current version</th>
        <th style="border-right:3px solid black">Target version</th>
        <th>Merged version (output)</th>
      </tr>
    </thead>
    <tbody valign="top">
    <tr>
      <td style="border-right:3px solid black">Keep user customizations on conflicts, but update sections with no conflicts</td>
      <td>
      <pre>
[{
    "framework":"MITRE ATT&CK",
    "tactic":{
      "id":"TA0003",
      "name":"Persistence",
      "reference":"https://attack.mitre.org/tactics/TA0003/"
    },
    "technique":[
      {
          "id":"T1098",
          "name":"Account Manipulation",
          "reference":"https://attack.mitre.org/techniques/T1098/"
      }
    ]
}]
        </pre>
      </td>
      <td>
        <pre>
[{
    "framework":"MITRE ATT&CK",
    "tactic":{
      "id":"TA0003",

>       "name":"My customized name",

Reviewer 3: Example here doesn't look correct. Users can only add and remove MITRE ATT&CK tacticts, techniques and subtechniques from the a specific MITRE ATT&CK version Kibana is aware of. `name` and `reference` fields are here only to simplify name retrival.

Author: Removing whole Mitre section for now.

The endpoint currently implemented contract needs to be updated as described in this [PoC](https://github.com/elastic/kibana/pull/144060).

### Changes to `/upgrade/_perform` endpoint

#### Changes to `/upgrade/_perform` endpoint contract

The endpoint's currently implemented contract needs to be updated as described in this [PoC](https://github.com/elastic/kibana/pull/144060).

### Changes to Rule Upgrade UX/UI flow

#### Bulk accepting upgrades with no conflicts

Making prebuilt rules customizable, plus the changes introduced to the `/upgrade/_review` update described above means that the Upgrade Review endpoint will now determine that some of the rules have `CONFLICT`s that cannot be automatically resolved (the current behaviour is updating all rules to their `target` version).

However, since we still want to give the users the ability to update all of their rules that have updates with no conflicts in an easy and fast way, we need to introduce a distinct UI that allows for this behaviour.

The current UI of the **Rule Updates** page has an button with the label "Update all" which currently simply updates all rules to their `target` version:

![](2024-02-23-14-09-16.png)

> We can replace this button for an **Update rules with no conflicts** button. Clicking this button would make a request to `/upgrade/_perform` with a payload that includes only rules that have no conflicts. That payload can be built out of the response of the `/upgrade/_review` endpoint, including only rules with no conflicts and specifying that all rules should be updated to their `MERGED` version, which represents the "clean" (conflictless) update.

Reviewer 3: Can `/upgrade/_perform` decide on its own which rules don't have conflicts and can be updated? In some rare cases it's possible a list of conflict free rules isn't relevant anymore since a new package has been just released but UI's state hasn't been updated.

Author: > Can /upgrade/_perform decide on its own which rules don't have conflicts and can be updated? 

No, the payload with conflict-free rules has to be built in the frontend, out of the response of `/upgrade/_review`. The pseudocode for that is below.

> In some rare cases it's possible a list of conflict free rules isn't relevant anymore since a new package has been just released but UI's state hasn't been updated.

That's true, but I wouldn't say that the rule aren't relevant anymore; such a request will simply update rules to a version that is one version less than the latest that is available. Once the user upgrades to this version, the rule will still appear in the Add Rules Page because a further update is still available.

Author: Two sides of the concurrency control mechanism in place:
- `version`: handles case of a new packagge being installed while the upgrade workflow is in progress (version in the payload is expected to be the last of the `security-rule` asset installed)
- `revision`: handles case of another user updating the rules while the user is doing the upgrade workflow

Author: - Create ticket for describing use case of "Updating all non-conflicting rules".

Rules whose diffing algorithm resulted in a `CONFLICT` need to be manually resolved or confirmed by the user via the UI before making the API request to upgrade the rule.

> See [Three-Way-Diff Component ticket](https://github.com/elastic/kibana/issues/171520) that details UI for solving conflicts.

> See [designs](https://www.figma.com/file/gLHm8LpTtSkAUQHrkG3RHU/%5B8.7%5D-%5BRules%5D-Rule-Immutability%2FCustomization?type=design&mode=design&t=LkauhLzUKUatF6cL-0#712870904).

## Other changes to UX/UI

### Rules Table page

- The Rules Table's filter should be updated so that rules can be fitlered according to their status as **custom**, **Elastic prebuilt** and **customized Elastic prebuilt**. Notice that **customized Elastic prebuilt** will be a subset of **Elastic prebuilt**.

### Rule Details page

- The Rule Details page shouls include some kind of indication or text field that lets the user know if the current rules is **custom**, **Elastic prebuilt** and **customized Elastic prebuilt**. 

## Design Discussion link

For an up-to-date discussion on the design changes needed for Milestone 3, please see [this ticket](https://github.com/elastic/kibana/issues/178211).

> 

Reviewer 4: We need to add Telemetry section - we currently collect info on the number of enabled disabled prebuilt and custom rules - potentially that one needs updates due to schema changes. @jpdjere

Reviewer 2: That's a great point 👍 

@approksiu Also, we have https://github.com/elastic/kibana/issues/140369 for telemetry that I think we should include into Milestone 3. Could you please review and update this ticket with requirements for Milestone 3?

Author: - Create ticket to address changes needed to existing Telemetry for number of enabled/disabled rules, updating to reflect changes in schema (no more `immutable` field, etc)

- [`createRules`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/create_rules.ts)
- [`patchRules`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/patch_rules.ts)
- [`updateRules`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/update_rules.ts)

This "overuse" of these 3 methods for a variety of user actions makes their logic tightly coupled and creates a considerable amount of complexity to CRUD functions that should remain logically simple.

For example: the `createRules` method is used in 3 use cases:
1. when creating custom rules with our Rule Creation endpoints
2. when importing rules, when the imported `rule_id` does not already exist in Kibana 
3. when upgrading rules, if a rule undergoes a `type` change, the existing rule is deleted a new one is created.

The same happens with `patchRules`. It is used:
1. when patching rules via our Rule Patch endpoints

> 2. when importing rules, when the imported `rule_id` is brand new

Author: Thanks :)

It's not in the export handler, it's in the importing handler. Specifically when you import a rule where the `rule_id` already exists

In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing two new fields: `prebuilt` and `customized`. Simultaneously, we need to deprecate the existing `immutable` field, which determines if a rule is an Elastic Prebuilt Rule and whether its fields can be modified/customized.

To ensure backward compatibility and avoid breaking changes, we will silently deprecate the `immutable` field but keep it unchanged (i.e. non-optional).

Since the migration of rules will be performed incrementally (see below), we have to assume that, at any given point in time, a user may have a mix of migrated and non-migrated rule saved objects. Consequently, we must continue supporting both versions while also carrying out the logic for migrating rules to the new schema.

In summary, the changes to be done to our rule schema are:

- Mark the `immutable` field as deprecated but keep unchanged for backwards compatibility.
    - Should remain a `boolean` type and non-optional (`required`)
- Add `prebuilt` field to the rule schema
    - `prebuilt` should have same type as `immutable`: `boolean`
    - Should be non-optional (`required`)

>     - During migration, set the `prebuilt` value to match the current value of `immutable`. This means: prebuilt rules should be migrated with a `prebuilt` value set to `true`, while custom rules should be set with a `prebuilt` value of `false`.

Reviewer 2: Would it be correct to say that on our API surface we will maintain the following invariant:

```ts
const alwaysTrue = rule.immutable === rule.prebuilt;
```

Author: Yes, clarified this in this section.

  threat: ThreatArray,
  to: RuleIntervalTo,
  references: RuleReferenceArray,
  version: RuleVersion,
  exceptionsList: RuleExceptionList.array(),
  relatedIntegrations: RelatedIntegrationArray.optional(),
  requiredFields: RequiredFieldArray.optional(),
  setup: SetupGuide.optional(),
});
```

```yaml
// common_attributes.schema.yaml
[...]
    IsRulePrebuilt:
      type: boolean
      description: Determines whether the rule is an Elastic Prebuilt Rule that can receive upstream updates via Fleet.

    IsPrebuiltRuleCustomized:
      type: boolean
      description: Determines whether the prebuilt rule has been customized by the user (i.e. any of its fields have been modified and diverged from the base value).
[...]
```

## Mapping changes

**Alert (rule objects) mapping**

No changes will be needed for the [mapping of rule saved objects](https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/common/saved_objects/rules/mappings.ts) (of type `alert`), since the new fields introduced will be part of the `params` field, which is a `flattened` field.

**Security Rules (prebuilt rule assets) mapping**

> No changes will be needed either for the `security-rule` [mapping](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_type.ts), our prebuilt rule assets. We currently have mappings for the `rule_id` and `version` fields, since we perform aggregations and filtering on those in the rule install and upgrade endpoints. No additional fields needs to be mapped in this phase.

Author: This is correct: my understanding as well is that the schema for `security-rule` assets shouldn't change (either for the `immutable` or the new `customized` and `prebuilt` rule fields).

This section refers to the ES mappings, so I'm going to add this point specifically for the Prebuilt Rule asset schema in the `## Necessary rule schema changes` section.

## Plan for carrying out migrations of rule SOs

### Migration strategy: perform migration on all write operations of the rules

> In order to migrate our users rule's data to the new schema, we should perform the migration of rules whenever any endpoint operation that writes/updates a rule saved object.

Author: > Let's provide readers of this RFC with context and explain why we have to migrate rules in our endpoints instead of using the migration mechanism.

Added this explanation in the `### Context` section above. I didn't write any migrations with the historical migration mechanism, so I have limited understanding of how it was carried out, but I think I described it more or less correctly. Can you take a look and tell me if anything I described there is wrong?

> Also, let's distinguish right from the beginning the two types of this migration:

Yes, makes sense. I'll rewrite this section with a clear split between them, and explaining them separately.

> In Elasticsearch, the type of rule saved objects is called `alert`.

This migration strategy of write/update operations means that a user's data will be migrated incrementally, and that both types of data (non-migrated and migrated saved objects) will coexist for an indeterminate amount of time. Therefore we have to maintain backwards data compatibility for the non-migrated data types.

The migration logic should take place as a first step within the handler logic of all endpoints that carry out write/update endpoint operations, and the endpoint should return the already-migrated rule(s).

### Rule Management endpoints that should include this migration logic are:

All of the following endpoints either fetch the rule before updating it, or send the rule's params as part of the body in the request. Therefore, in order to carry out the migration, we can check if the `params` fields already contains the `prebuilt` and `customized` fields, create them within the `params` if they don't, and then proceed to saving them to Elasticsearch.

- **Update Rule** - `PUT /rules`: same as Patch Rule.

>     - Used **only** when updating/modifying a single rule via the Rule Details page

Author: Yes, thanks. Fixed.

>     - Used **only** when updating/modifying a single rule via the Rule Details page

Author: I have that below in the section:
### Technical implementation of migration
I will address your comments for that in that section

- **Patch Rule** - `PATCH /rules`: should migrate patched rule to prevent the endpoint from returning an outdated (non-migrated) rule.
    - Used for attaching shared exceptions list to rules
- **Bulk Update Rules** - `PUT /rules/_bulk_update`: same as for Patch Rule
    - Deprecated and unused by the UI (might still be used by API users)
- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`: same as for Patch Rule
    - Deprecated and unused by the UI (might still be used by API users)
- **Import Rules** - `POST /rules/_import`: 

>     - if the user imports rules with the updated rule schema (`prebuilt` and `customized` fields are present), then save rule as-is. 

Reviewer 2: If only `prebuilt` and `customized` fields are present, wouldn't we need to initialize the `immutable` field if it wasn't provided by the user?

Author: This is true, thanks. Added a detailed explanation for this case.

>     - if the user imports a "legacy" rule (with the `immutable` field, and no `prebuilt` or `customized` fields), the `immutable` field will be saved for backwards compatibility as well, and the `prebuilt` and `customized` fields should always be set to `false` (importing a non-migrated rule means that the rule could not be modified or exported in the Kibana version where it was exported from, so it could not have been a prebuilt rule).

Reviewer 2: > if the user imports a "legacy" rule (with the `immutable` field, and no `prebuilt` or `customized` fields)

Does this refer to the `immutable: false` case? Basically, custom rules that our users can export and import right now. Are we talking about backward compatibility with ndjson files containing such rules?

Also, not sure I understand this sentence, could you rephrase?

> importing a non-migrated rule means that the rule could not be modified or exported in the Kibana version where it was exported from, so it could not have been a prebuilt rule

Author: > Does this refer to the immutable: false case? Basically, custom rules that our users can export and import right now.

Not only that case, but also rules with `immutable: true` case and non-migrated (`prebuilt` and `customized` do not exist).

> Are we talking about backward compatibility with ndjson files containing such rules?

Yes: we want to handle the case of import of both migrated and non-migrated rules via an ndjson file. I understand that importing rules with `immutable: false` "should" be the only case for non-migrated rules (since `immutable: true` rules cannot be exported), but I think we should also handle the case of users manually modifying the values of rules in the NDJSON file. 
I think that, if a user attempts to import a rule with  `immutable: true` and missing `prebuilt` and `customized` rules, then the most likely scenario is a user trying to import Elastic prebuilt rules from a modfied NDJSON file and we should mark those rules as `prebuilt: true` in ES and attempt to track them to rules in the Fleet package via their `rule_id`.

> Also, not sure I understand this sentence, could you rephrase?

I'm removing that phrase, I thought this over again and the paragraph above makes the most sense to me. I'm writing that idea in the RFC.

Author: I moved everything related to them import endpoint to the [Importing rules](#importing-rules){#toc-importing-rules} section and linking there.

- **Bulk Actions** - `POST /rules/_bulk_action`: 
    - This endpoint includes many types of actions, and only a subset of them should perform a migration.

>     - This endpoint also includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`.

Reviewer 2: > No migration logic should take place for dry run requests, i.e when `dry_run=true`.

Why the logic should be different depending on this parameter's value? Shouldn't it only affect the fact whether we save these changes or not?

Author: Because when the `dry_run` parameter is set to `true`, none of the the logic paths of the current endpoint code save changes to ES.

I'm adding this to the sentence, for clarity.

    - Actions that should perform migration logic:
        - Edit:

>             - endpoint should migrate rule while performing migration. We can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed.

Reviewer 2: ```suggestion
            - endpoint should migrate rule while applying bulk editing actions. We can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed.
```

    - Actions that should **not** perform migration logic:
        - Enable
        - Disable
        - Delete
        - Export:

>             - export endpoint logic does not write/modify rules, only reads them and makes them available for download. Adding new writing logic could have a big impact on its performance.

Reviewer 2: It should still normalize rules on-read before returning them in the ndjson format, no?

Author: Yes, I'm adding this in a separate section detailing on-read normalization.

>             - we will now allow the export of prebuilt rules

Reviewer 2: This needs its own section and explanation what exactly we will support and how it can work.

            - export endpoint should therefore only export the rule in their current state (migrated or non-migrated), and the import endpoint will take care of the migration logic
        - Duplicate:
            - this action should not migrate the original rule being duplicated, as it would violate the principle of least surprise
            - the new rule being created should have the new schema

>             - we should only update the duplication logic to account for possibility that the new `prebuilt` and `customized` fields already exist in the original rule

Reviewer 2: Nit: We will need to handle this case in pretty much every endpoint.

Author: Yes, removing this from here, it will be clarified with code in the implementation details of this endpoint.

            - per definition: if a prebuilt rule is duplicated, we need to set the new rule's `prebuilt` value to `false` (and keep the current logic that also sets `immutable` to `false`).

**NOTE:** Endpoints that depend on the Rules Client's methods need to modify the rules' params (prebuilt and customized) before the Rule Client methods call the`validateRuleTypeParams`, because that util validates the passed parameters using the validators defined in `x-pack/plugins/security_solution/server/lib/detection_engine/rule_types`. Since those are an intersection between `BaseRulesParams` and the specific rule type's fields, `prebuilt` and `customized` are non-optional, which would cause the util to reject the validation of the params for non-migrated rules.

> ### Other Rule Management endpoints that will need to be adapted to new schema, but will not need migration logic:

Reviewer 2: Meaning migration on-write logic?

Author: Yes, it's clearer now.

- **Create Rules** - `POST /rules`:

>     - If the request includes either the `immutable` and `prebuilt` fields, ignore both parameters and set both fields to `false`. This is analogous to the current logic: we don't want users to be able to create prebuilt rules, only install them and customize them, and always be able to compare them to a `security-rule` asset.

Reviewer 2: Currently, we don't support the `immutable` field in any of the endpoints' request parameters. We shouldn't support the `prebuilt` and `customized` fields either, because their values should be controlled by the app on the server side, and never controlled by users.

Specifically, this endpoint should be able to create _only_ rules with:

```
immutable: false,
prebuilt: false,
customized: false
```

I guess this is what you meant by saying "ignore both parameters and set both fields to `false`", but I think it should be explicitly called out that all the three fields shouldn't be supported in request parameters of all our endpoints, except maybe the rules import endpoint.

Author: Thanks, I added and clarified this.

- **Get Prebuilt Rules Status** - `GET /prebuilt_rules/status` (Internal): Uses `IPrebuiltRuleObjectsClient` to retrieve instances of prebuilt rules according to the `immutable` field. Adapt logic to rely on new `alert.attributes.params.prebuilt` field, with fallback to the original, for backwards compatibility.
- **(LEGACY) Install Prebuilt Rules And Timelines** - `PUT /rules/prepackaged`: currently depends on rules `alert.attributes.params.immutable` to fetch custom rules and prebuilt rules. Adapt logic to rely on new `alert.attributes.params.prebuilt` field, with fallback to the original `immutable`, for backwards compatibility.
- **Installation and Upgrade `_review` and `_perform` endpoints:**
    - All four endpoints use `IPrebuiltRuleObjectsClient` to retrieve instances of prebuilt rules according to the `immutable` field. Adapt logic to rely on new `alert.attributes.params.prebuilt` field, with fallback to the original, for backwards compatibility.
        - **Review Rule Installation** - `POST /prebuilt_rules/installation/_review` (Internal):

>             - Uses `convertPrebuiltRuleAssetToRuleResponse`, where we need to set new `prebuilt` field to `true` and `customized` to `false`.

Reviewer 2: What about the `immutable` field?

        - **Perform Rule Installation** - `POST /prebuilt_rules/installation/_install` (Internal):

>             - When new prebuilt rules are created, pass in new `prebuilt` param set to `true`.

Reviewer 2: What about the `immutable` and `customized` params?

        - **Review Rule Upgrade** - `POST /prebuilt_rules/upgrade/_review` (Internal)

>             - Uses `convertPrebuiltRuleAssetToRuleResponse`, where we need to set new `prebuilt` field to `true` and `customized` to `false`.

Reviewer 2: What about the `immutable` field?

            - This endpoint will need further changes, detailed further down, out of the scope of SO and rule schema updates.
        - **Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal)

>             - Should update the `upgradeRule` util to include new `prebuilt` and `customized` fields.

Reviewer 2: Missing an explanation how and based on what these endpoints would calculate the value of the `customized` field. Mind that there can be cases where non-customized prebuilt rules can become customized after an upgrade, by the upgrade itself, if the user changes the values of some fields manually on upgrade.

Author: Will make this explanation where I detail the changes needed for this endpoint for allowing customized fields, as they are strongly related. Left a note about that.

- **Get Space Health** - `GET or POST /detection_engine/health/_space` (internal): same as Get Cluster Health, uses `RuleObjectsHealthClient` internally and aggregates over `alert.attributes.params.immutable`.
- **Get Rule Health** - `POST /detection_engine/health/_rule` (internal): does not depend on any of the affected params, i.e. no changes are needed here.

#### Rule Execution Logs

Rule execution logging is not affected in any way depending on whether the rules is prebuilt or not, so any changes to the rule schema won't affect these endpoints.

<details>
  <summary> <b>Endpoints that will not need any changes</b></summary>
<ul>
    <li>Find Rules - <code>GET /rules/_find</code>: Does not depend on any fixed existing params, although filtering by rule params can be passed to it in the request query params.</li>
    <li>Read Rule - <code>GET /rules</code>: similar to "find rules", but for a single rule, retrieved by <code>id</code> or <code>rule_id</code></li>
    <li>Delete Rules - <code>DELETE /rules</code></li>

>     <li>Bulk Delete Rules - <code>DELETE or POST /rules/_bulk_delete</code></li>

Author: Yes, detailed.

    <li>Read Tags - <code>GET /tags</code></li>
</ul>
</details>

### Technical implementation of migration

> The migration of the rule saved objects can be implemented directly in the route handlers of the endpoints mentioned above. (The examples below are just implementation examples; we should most probably encapsulate the logic in a reusable function and import it into wherever we want to perform the migration)

Author: Yes, I'm creating reusable migrations function that will cover all needed migration cases.

          checkDefaultRuleExceptionListReferences({ exceptionLists: request.body.exceptions_list });

          await validateRuleDefaultExceptionList({
            exceptionsList: request.body.exceptions_list,
            rulesClient,
            ruleRuleId: request.body.rule_id,
            ruleId: request.body.id,
          });

          const existingRule = await readRules({
            rulesClient,
            ruleId: request.body.rule_id,
            id: request.body.id,
          });

          await validateResponseActionsPermissions(
            ctx.securitySolution,
            request.body,
            existingRule
          );

          /* New rule schema migration logic
          *
          * If a rule was successfully retrieved, and its `prebuilt` param doesn't exist
          * modify its params to add the prebuilt and customized fields
          */
          if (existingRule && existingRule?.params.prebuilt === undefined) {
            existingRule.params = {
              ...existingRule?.params,
              prebuilt: existingRule?.params.immutable,
              customized: existingRule?.params.customized ?? false,
            };

>           }

Author: Will do in the reusable migration functions.

          const rule = await updateRules({
            rulesClient,
            existingRule,
            ruleUpdate: request.body,
          });

          if (rule != null) {
            return response.ok({
              body: transformValidate(rule),
            });
          } else {
            const error = getIdError({ id: request.body.id, ruleId: request.body.rule_id });
            return siemResponse.error({
              body: error.message,
              statusCode: error.statusCode,
            });
          }
        } catch (err) {
          const error = transformError(err);
          return siemResponse.error({
            body: error.message,
            statusCode: error.statusCode,
          });
        }
      }
    );
};

> ```

Author: Done

```ts
/**
 * takes list of bulkEdit actions and apply them to rule.params by mutating it
 * @param existingRuleParams
 * @param actions
 * @returns mutated params, isParamsUpdateSkipped flag
 */
export const ruleParamsModifier = (
  existingRuleParams: RuleAlertType['params'],
  actions: BulkActionEditForRuleParams[]
): RuleParamsModifierResult<RuleAlertType['params']> => {
  let isParamsUpdateSkipped = true;

  const modifiedParams = actions.reduce((acc, action) => {
    const { ruleParams, isActionSkipped } = applyBulkActionEditToRuleParams(acc, action);

    // The rule was updated with at least one action, so mark our rule as updated
    if (!isActionSkipped) {
      isParamsUpdateSkipped = false;
    }
    return { ...acc, ...ruleParams } as RuleAlertType['params'];
  }, existingRuleParams);

  /* New rule schema migration logic
  *
  * If a rule's `prebuilt` param doesn't exist
  * modify its params to add the prebuilt and customized fields
  */
  if (modifiedParams.prebuilt === undefined) {
    modifiedParams.prebuilt = modifiedParams.immutable;
    modifiedParams.customized = modifiedParams.customized ?? false;

>   }

Author: Explained how this will be in `migratePrebuiltSchemaOnRuleBulkEdit` section

From [`x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts`](x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts):

```ts
/* If `rule_id` does not exist, create a new rule */
if (rule == null) {

  /* Set the prebuilt param to the parsed `prebuilt` param, or `immutable` for non-migrated rules 
   * If neither are defined, `createRules` defaults its `prebuilt` param to `false`.
   */
  const isNewRulePrebuilt = parsedRule.prebuilt || parsedRule.immutable;

  /* If rule is not migrated, set `customized` to false. Otherwise use the input value */
  const isNewRuleCustomized = parsedRule.customized === undefined ? false : parsedRule.customized;

  await createRules({
    rulesClient,
    params: {
      ...parsedRule,
      customized:  isNewRuleCustomized,
      exceptions_list: [...exceptions],
    },

>     prebuilt: isNewRulePrebuilt,

Author: The `createRules` function is like this because the parameters passed in the `CreateRuleOptions` object that it takes as parameter extend the `RuleCreateProps`, which don't have `prebuilt` or `immutable` as a field within it:

https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/create_rules.ts

```ts
export interface CreateRulesOptions<T extends RuleCreateProps = RuleCreateProps> {
  rulesClient: RulesClient;
  params: T;
  id?: string;
  prebuilt?: boolean;
  // immutable?: boolean;
  defaultEnabled?: boolean;
  allowMissingConnectorSecrets?: boolean;
}

export const createRules = async ({
  rulesClient,
  params,
  id,
  // immutable = false,
  prebuilt = false,
  defaultEnabled = true,
  allowMissingConnectorSecrets,
}: CreateRulesOptions): Promise<SanitizedRule<RuleParams>> => {
  // const internalRule = convertCreateAPIToInternalSchema(params, immutable, defaultEnabled);
  const internalRule = convertCreateAPIToInternalSchema(params, prebuilt, defaultEnabled);
```

Since `createRules` is used to create new custom rules, create prebuilt rules (and also create rules when importing them), the `immutable` (or `prebuilt` field) has to be a separate parameter to the function, that is sometimes passed as `true`, and sometimes as `false` (and should set the value for both `immutable` and `prebuilt` in the future).

`customized` should be always set to `false` since we are creating a new rule.

>     prebuilt: isNewRulePrebuilt,

Author: I added this explanation in the RFC where I explain the changes needed to the `createRules` method.

>     prebuilt: isNewRulePrebuilt,

Author: Sorry, one thing I missed here but is even more important: the `params` object here passed to `createRule` extends from `RuleCreateProps`, which itself doesn't contain the `immutable` or `prebuilt` fields.

This makes sense: we don't want to have these parameters as part of the Rule creation API endpoint.

      };
  }

  await patchRules({
    rulesClient,
    existingRule: rule,
    nextParams: {
      ...migratedParams,
      exceptions_list: [...exceptions],
    },
    allowMissingConnectorSecrets,
    shouldIncrementRevision: false,
  });
  resolve({
    rule_id: parsedRule.rule_id,
    status_code: 200,
  });
}
```

$\space$

## Exporting and importing rules

**Current behaviour:**

- Custom Rules can be exported and imported
- When trying to import a custom rule, its `rule_id` should not already exist in any of the installed rules. Otherwise the import logic will fail : `rule_id: {rule_id} already exists`
- Prebuilt Rules cannot be exported
- Prebuilt Rules cannot be imported (if you try to import a rule that has been set the `immutable` field to `true` manually, the import endpoint will reject it as the `immutable` field is expected to be `false`)

**Updated behaviour**

We will now allow the user to export and import both **prebuilt** and **custom rules**.

### Exporting rules

The user will now be able to export both custom and prebuilt rules (including any customizations that might have been performed on its modifiable fields).

> The export endpoints (bulk export and bulk action in export mode) will not carry out any migration logic and will export the rule as-is. Since these endpoints only read the rules and create an `ndjson` file to make them available for export, we don't want to add any additional writing/patching logic to migrate the rule's schema on Elasticsearch.

Reviewer 2: > The export endpoints (bulk export and bulk action in export mode) will not carry out any migration logic and will export the rule as-is

Can you elaborate why? I think the export endpoint should normalize rule saved objects to the new schema ("migrate on-read") before returning them from the export endpoints, just like any other "GET" endpoints that return rules.

Might be that there's a confusion here around physical migration on-write (with saving changes to ES) vs virtual normalization on-read (w/o saving changes to ES). Commented on this above in some comment. Let's clearly distinguish these two types of transformations in the RFC and use different terms for them.
